{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as pyplot\n",
    "from math import log, sqrt\n",
    "from os.path import join as joinpath\n",
    "\n",
    "from aminoacid import AminoAcid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project 4 - Secondary structure prediction\n",
    "\n",
    "Proteins have four levels of structure:\n",
    "* The **primary** structure is the sequence of amino acids that makes up the protein.\n",
    "* The **secondary** structure refers to particular shapes that sub-sequences of the protein tend to form, due to *hydrogen bonds*. The most common among these are *alpha helices* and *beta sheets*.\n",
    "* The **tertiary** structure is how the whole protein is \"folded\" (i.e. its 3D structure). The folding is due to hydrophobic interactions, and stops when the shape is stabilized by other interactions.\n",
    "* The **quartenary** structure is particular to *multimers*, proteins that are made up of multiple subunits.  It describes how these subunits are arranged together.\n",
    "\n",
    "The goal of this project is to predict the secondary structure of a protein, based on its primary structure. This is useful in the context of multiple sequence alignment, since proteins that exerce the same function are likely to have similar secondary structure as well as related primary structures. Fortunately for us, secondary structures can be observed experimentally via multiple techniques, granting us the possibility to train and verify our prediction system with real-world data. Furthermore, the folding of proteins (into their secondary and tertiary stable structures) is highly deterministic, which means it can be predicted based on the primary structure alone.\n",
    "\n",
    "## DSSP definition\n",
    "\n",
    "DSSP stands for *Define Secondary Structure of Proteins* and is a standard for how the atomic 3D arrangement of a protein is translated into secondary structures. DSSP admits eight types of secondary structures and assigns one to each amino acid from a protein by examining their spacial coordinates. We won't be implementing DSSP, however we will need to parse `.dssp` files in order to extract secondary structure information to train and verify our prediction system. Here is a class that parses such files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class DSSP:\n",
    "\tdef __init__(self, filePath):\n",
    "\t\t\n",
    "\t\t#Interesting columns : (start index, end index)\n",
    "\t\tself.columns = (\"RESIDUE\", \"AA\", \"STRUCTURE\")\n",
    "\t\tself.residues = []\n",
    "\t\t\n",
    "\t\t#Metadata\n",
    "\t\tself.identifier = \"\"\n",
    "\t\tself.protein = \"\"\n",
    "\t\tself.organism = \"\"\n",
    "\t\t\n",
    "\t\t#Parsing\n",
    "\t\twith open(filePath, 'r') as dsspFile:\n",
    "\t\t\tcolumnIndex = {col : (0, 0) for col in self.columns}\n",
    "\t\t\tlineIsData = False\n",
    "\t\t\t\n",
    "\t\t\tfor line in dsspFile.readlines():\n",
    "\t\t\t\t\n",
    "\t\t\t\tif lineIsData:\n",
    "\t\t\t\t\tdata = []\n",
    "\t\t\t\t\tfor column in self.columns:\n",
    "\t\t\t\t\t\tstart, end = columnIndex[column]\n",
    "\t\t\t\t\t\tdata.append(line[start:end])\n",
    "\t\t\t\t\t\n",
    "\t\t\t\t\tself.residues.append(data)\n",
    "\t\t\t\t\t\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tif line.strip()[0] == \"#\":\n",
    "\t\t\t\t\t\tlineIsData = True\n",
    "\t\t\t\t\t\tfor column in self.columns:\n",
    "\t\t\t\t\t\t\tstartIndex = line.find(column)\n",
    "\t\t\t\t\t\t\tendIndex = startIndex + len(column)\n",
    "\t\t\t\t\t\t\tendIndex = endIndex + (len(line[endIndex:]) - len(line[endIndex:].lstrip())) - 1\n",
    "\t\t\t\t\t\t\tcolumnIndex[column] = (startIndex, endIndex)\n",
    "\t\t\t\t\telif line.startswith(\"HEADER\"):\n",
    "\t\t\t\t\t\tself.identifier = line.split()[-2]\n",
    "\t\t\t\t\t\t\n",
    "\t\t\t\t\telif line.startswith(\"COMPND\"):\n",
    "\t\t\t\t\t\tself.protein = line.split(\":\")[1].split(\";\")[0].strip()\n",
    "\t\t\t\t\t\tself.protein = \" \".join(self.protein.split())\n",
    "\t\t\t\t\t\t\n",
    "\t\t\t\t\telif line.startswith(\"SOURCE\"):\n",
    "\t\t\t\t\t\tself.organism = line.split(\":\")[1].split(\";\")[0].strip()\n",
    "\t\t\t\t\t\tself.organism = \" \".join(self.organism.split())\n",
    "\t\t\t\t\t\t\n",
    "\t\n",
    "\tdef __repr__(self):\n",
    "\t\tres = []\n",
    "\t\tfor values in self.residues:\n",
    "\t\t\tres.append(str(values))\n",
    "\t\treturn \"\\n\".join(res)\n",
    "\t\t\n",
    "\t\n",
    "\tdef getSequenceStructure(self, chain):\n",
    "\t\tstructs = {\"H\":\"H\",\"G\":\"H\",\"I\":\"H\",\"E\":\"E\",\"B\":\"E\",\"T\":\"T\",\"C\":\"C\",\"S\":\"C\",\" \":\"C\"}\n",
    "\t\tsequence = []\n",
    "\t\tstructure = []\n",
    "\t\t\n",
    "\t\tfor residue in self.residues:\n",
    "\t\t\tif residue[0][-1] == chain:\n",
    "\t\t\t\tsequence.append(residue[1][0])\n",
    "\t\t\t\tstructure.append(structs[residue[2][0]])\n",
    "\t\t\n",
    "\t\treturn \"\".join(sequence), \"\".join(structure)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that we won't be using all eight structures, but rather regroup them into four classes:\n",
    "* **Helix** (H) regroups 3,4 and 5-turn helixes\n",
    "* **Sheet** (E) regroups parallel/antiparallel $\\beta$-sheets and isolated $\\beta$-bridges\n",
    "* **Turn** (T) is the hydrogen bonded turn\n",
    "* **Coil** (C) regroups coils (no structure) and bends\n",
    "\n",
    "\n",
    "## GOR prediction\n",
    "\n",
    "GOR stands for *Garnier-Osguthorpe-Robson* and is a secondary structure prediction method based on information theory. It has had several releases, each increasing the prediction accuracy, but we will only focus on the GOR III version here. This version uses two kinds of information to issue a prediction, all based on known protein-structure pairs parsed from a *training dataset*. In the following formulas, $R_j$ is the residue (amino acid) at index $j$ whose structure is being predicted, $S_j$ is one of the structures, $n-S_j$ represents all of the structures except for $S_j$, $f_{c_1,...c_k}$ is the frequency with which all conditions $c_1$ through $c_k$ are met within the training dataset and $I(\\Delta S, ...) = I(S, ...) - I(n-S, ...)$ is the information difference between the predictions concerning $S$ and $n-S$.\n",
    "* **Individual** information concerns only the amino acid at position $j$: $$I(\\Delta S_j, R_j) = \\log{\\left( \\frac{f_{S_j,R_j}}{f_{n-S_j,R_j}} \\right)} + \\log{\\left( \\frac{f_{n-S_j}}{f_{S_j}} \\right)}$$\n",
    "* **Directional** information was introduced in version 2 and concerns the amino acids surrounding position $j$, from $j-n$ to $j+n$: $$I(\\Delta S_j, R_{j+m}) = \\log{\\left( \\frac{f_{S_j,R_{j+m}}}{f_{n-S_j,R_{j+m}}} \\right)} + \\log{\\left( \\frac{f_{n-S_j}}{f_{S_j}} \\right)}$$\n",
    "* **Pair-wise** information has replaced directional information since version 3 and concerns the pairs $(R_j, R_{j+m}) \\forall m \\in [-n, -1] \\cup [1, n]$: $$I(\\Delta S_j, R_{j+m} | R_j) = \\log{\\left( \\frac{f_{S_j,R_{j+m},R_j}}{f_{n-S_j,R_{j+m},R_j}} \\right)} + \\log{\\left( \\frac{f_{n-S},R_j}{f_{S},R_j} \\right)}$$\n",
    "\n",
    "Overall, the formula applied for the GOR III prediction is: $$I(\\Delta S_j, R_{j-n} ... R_{j+n}) = I(\\Delta S_j, R_j) + \\sum_{m=-n, m \\neq 0}^{m=n}{I(\\Delta S_j, R_{j+m} | R_j)}$$\n",
    "\n",
    "Here is an implementation of the algorithm, that we can train with new sequences then use to predict the structure of other sequences:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "class GOR3:\n",
    "\t\"\"\"\n",
    "\tImplements the GOR III secondary structure prediction algorithm.\n",
    "\tObjects of this class must be trained with known examples of sequences and their structure\n",
    "\tbefore being able to predict the structures of new sequences.\n",
    "\t\"\"\"\n",
    "\n",
    "\tdef __init__(self):\n",
    "\t\tself.structures = \"HETC\"\n",
    "\t\tself.roc = {(s, r):0 for s in self.structures for r in (\"TP\", \"TN\", \"FP\", \"FN\")}\n",
    "\t\tself.thresholdRange = np.arange(0,1,0.1)\n",
    "\t\tself.thresholdROC = {(s, r, t):0 for s in self.structures for r in (\"TP\", \"TN\", \"FP\", \"FN\") for t in self.thresholdRange}\n",
    "\t\tself.correctPred = 0\n",
    "\t\tself.totalPred = 0\n",
    "\t\tself.neighbourOffset = 8\n",
    "\t\t\n",
    "\t\tself.trainings = 0 #Number of trainings (one per AA)\n",
    "\t\tself.strucCount = {s:0 for s in self.structures}\n",
    "\t\tself.pairCount = {(s, a):0 for s in self.structures for a in AminoAcid.getAllNames()}\n",
    "\t\tself.tripletCount = {}\n",
    "\t\tfor s in self.structures:\n",
    "\t\t\tfor a in AminoAcid.getAllNames():\n",
    "\t\t\t\tfor na in AminoAcid.getAllNames(): #Neighbour AA\n",
    "\t\t\t\t\tself.tripletCount[(s, a, na)] = 0\n",
    "\t\t\n",
    "\t\t\n",
    "\tdef train(self, sequence, structure):\n",
    "\t\t\"\"\"\n",
    "\t\tTrains the system with a known example of a sequence and its structure.\n",
    "\t\t\"\"\"\n",
    "\t\tself.trainings += len(sequence)\n",
    "\t\t\n",
    "\t\tfor index in range(len(sequence)):\n",
    "\t\t\tcurAminoacid = sequence[index]\n",
    "\t\t\tcurStructure = structure[index]\n",
    "\t\t\tself.strucCount[curStructure] += 1\n",
    "\t\t\tself.pairCount[(curStructure, curAminoacid)] += 1\n",
    "\t\t\t\n",
    "\t\t\tfor neiAminoacid in self.neighbourValues(sequence, index):\n",
    "\t\t\t\tself.tripletCount[(curStructure, curAminoacid, neiAminoacid)] += 1\n",
    "\t\n",
    "\t\n",
    "\tdef predict(self, sequence, realStructure=None):\n",
    "\t\t\"\"\"\n",
    "\t\tReturns the predicted structure of 'sequence', based on received training.\n",
    "\t\t\"\"\"\n",
    "\t\tstructure = [] #Result: predicted structure\n",
    "\t\t\n",
    "\t\t#Predict structures for each aminoacid in sequence\n",
    "\t\tfor index in range(len(sequence)):\n",
    "\t\t\tcurAminoacid = sequence[index]\n",
    "\t\t\t\n",
    "\t\t\t#First possible structure\n",
    "\t\t\tpredStructure = self.structures[0]\n",
    "\t\t\tpredScore = self.__getScore(sequence, index, predStructure)\n",
    "\t\t\t\n",
    "\t\t\t#Other structures\n",
    "\t\t\tfor curStructure in self.structures[1:]:\n",
    "\t\t\t\tcurScore = self.__getScore(sequence, index, curStructure)\n",
    "\t\t\t\t\n",
    "\t\t\t\t#Remember structure that gives best score\n",
    "\t\t\t\tif curScore > predScore:\n",
    "\t\t\t\t\tpredStructure = curStructure\n",
    "\t\t\t\t\tpredScore = curScore\n",
    "\t\t\t\t\t\n",
    "\t\t\tstructure.append(predStructure)\n",
    "\t\tstructure = \"\".join(structure)\n",
    "\t\t\n",
    "\t\tif not realStructure is None:\n",
    "\t\t\tself.__quality(sequence, structure, realStructure)\n",
    "\t\t\n",
    "\t\treturn structure\n",
    "\t\t\n",
    "\t\n",
    "\tdef __getScore(self, sequence, index, struct):\n",
    "\t\t\"\"\"\n",
    "\t\tReturns I(deltaS, R) as defined by the GOR III algorithm.\n",
    "\t\t\"\"\"\n",
    "\t\taminoacid = sequence[index]\n",
    "\t\tscoreTerms = []\n",
    "\t\t\n",
    "\t\tscore = self.pairCount[(struct, aminoacid)]\n",
    "\t\tscore /= sum(self.pairCount[(otherStruct, aminoacid)] for otherStruct in self.getStructures(struct))\n",
    "\t\tscoreTerms.append(score)\n",
    "\t\t\n",
    "\t\tscore = (self.trainings - self.strucCount[struct]) / self.strucCount[struct]\n",
    "\t\tscoreTerms.append(score)\n",
    "\t\t\n",
    "\t\tfor neiAminoacid in self.neighbourValues(sequence, index):\n",
    "\t\t\tscore = self.tripletCount[(struct, aminoacid, neiAminoacid)]\n",
    "\t\t\tscore /= sum(self.tripletCount[(otherStruct, aminoacid, neiAminoacid)] for otherStruct in self.getStructures(struct))\n",
    "\t\t\tscoreTerms.append(score)\n",
    "\t\t\t\n",
    "\t\t\tscore = sum(self.pairCount[(otherStruct, aminoacid)] for otherStruct in self.getStructures(struct))\n",
    "\t\t\tscore /= self.pairCount[(struct, aminoacid)]\n",
    "\t\t\tscoreTerms.append(score)\n",
    "\t\t\t\n",
    "\t\treturn sum(log(s) for s in scoreTerms)\n",
    "\t\t\t\n",
    "\t\t\t\n",
    "\t\t\n",
    "\tdef neighbourOffsets(self):\n",
    "\t\tfor offset in range(-self.neighbourOffset, self.neighbourOffset+1):\n",
    "\t\t\tif offset!=0:\n",
    "\t\t\t\tyield offset\n",
    "\t\n",
    "\t\n",
    "\tdef neighbourValues(self, sequence, index):\n",
    "\t\tfor offset in self.neighbourOffsets():\n",
    "\t\t\tneiIndex = index + offset\n",
    "\t\t\tif neiIndex >= 0 and neiIndex < len(sequence):\n",
    "\t\t\t\tyield sequence[neiIndex]\n",
    "\t\n",
    "\t\n",
    "\tdef getStructures(self, exclude=None):\n",
    "\t\tfor s in self.structures:\n",
    "\t\t\tif s != exclude:\n",
    "\t\t\t\tyield s\n",
    "\t\n",
    "\t\n",
    "\tdef __quality(self, sequence, structure, reality):\n",
    "\t\tself.totalPred += len(structure) #Total predictions\n",
    "\t\tself.correctPred += sum([1 if s==r else 0 for s,r in zip(structure, reality)]) #Correct predictions\n",
    "\t\t\n",
    "\t\tfor classifier in self.structures:\n",
    "\t\t\t\n",
    "\t\t\tfor threshold in self.thresholdRange:\n",
    "\t\t\t\tfor index in range(len(sequence)):\n",
    "\t\t\t\t\tscore = self.__getScore(sequence, index, classifier)\n",
    "\t\t\t\t\treal = reality[index]\n",
    "\t\t\t\t\tif score >= threshold: #Positive\n",
    "\t\t\t\t\t\tif classifier == real: #True\n",
    "\t\t\t\t\t\t\tself.roc[(classifier, \"TP\")] += 1\n",
    "\t\t\t\t\t\t\tself.thresholdROC[(classifier, \"TP\", threshold)] += 1\n",
    "\t\t\t\t\t\telse: #False\n",
    "\t\t\t\t\t\t\tself.roc[(classifier, \"FP\")] += 1\n",
    "\t\t\t\t\t\t\tself.thresholdROC[(classifier, \"FP\", threshold)] += 1\n",
    "\t\t\t\t\telse: #Negative\n",
    "\t\t\t\t\t\tif classifier != real: #True\n",
    "\t\t\t\t\t\t\tself.roc[(classifier, \"TN\")] += 1\n",
    "\t\t\t\t\t\t\tself.thresholdROC[(classifier, \"TN\", threshold)] += 1\n",
    "\t\t\t\t\t\telse: #False\n",
    "\t\t\t\t\t\t\tself.roc[(classifier, \"FN\")] += 1\n",
    "\t\t\t\t\t\t\tself.thresholdROC[(classifier, \"FN\", threshold)] += 1\n",
    "\t\n",
    "\t\n",
    "\tdef getQuality(self):\n",
    "\t\ttp = sum([self.roc[(c, \"TP\")] for c in self.structures])\n",
    "\t\ttn = sum([self.roc[(c, \"TN\")] for c in self.structures])\n",
    "\t\tfp = sum([self.roc[(c, \"FP\")] for c in self.structures])\n",
    "\t\tfn = sum([self.roc[(c, \"FN\")] for c in self.structures])\n",
    "\t\tmcc = (tp*tn - fp*fn) / sqrt((tp+fp)*(tp+fn)*(tn+fp)*(tn+fn))\n",
    "\t\tq3 = self.correctPred / self.totalPred\n",
    "\t\treturn q3, mcc\n",
    "\t\n",
    "\t\n",
    "\tdef plotROC(self):\n",
    "\t\tfor classifier in self.structures:\n",
    "\t\t\tx, y = [], []\n",
    "\t\t\tfor threshold in self.thresholdRange:\n",
    "\t\t\t\ttp = sum([self.roc[(c, \"TP\")] for c in self.structures])\n",
    "\t\t\t\ttn = sum([self.roc[(c, \"TN\")] for c in self.structures])\n",
    "\t\t\t\tfp = sum([self.roc[(c, \"FP\")] for c in self.structures])\n",
    "\t\t\t\tfn = sum([self.roc[(c, \"FN\")] for c in self.structures])\n",
    "\t\t\t\ttpr = tp / (tp + fn)\n",
    "\t\t\t\ttnr = tn / (tn + fp)\n",
    "\t\t\t\tx.append(tpr)\n",
    "\t\t\t\ty.append(tnr)\n",
    "\t\t\tpyplot.plot(x,y)\n",
    "\t\t\tpyplot.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prediction quality\n",
    "\n",
    "stuff about Q3, MCC, ROC curves"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results\n",
    "\n",
    "Time to run that bytecode.\n",
    "\n",
    "### Parsing DSSP files\n",
    "\n",
    "The first step towards predicting structures is to gather training and validation data. Fortunately, our dear TA has provided us with all the `.dssp` files we need. The only miscellaneous stuff here is that we're not interested in the whole data, just some chains from some files (specified in `info` files). The output goes in the `PARSED` files in a format similar to `fasta`, where the structure follows the sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open(joinpath(\"dataset\", \"CATH_info.txt\")) as infoFile:\n",
    "\twith open(joinpath(\"dataset\", \"CATH_info-PARSED.txt\"), 'w') as outFile:\n",
    "\t\tdsspPath = joinpath(\"dataset\", \"dssp\", \"\")\n",
    "\t\tfor line in infoFile.readlines():\n",
    "\t\t\td = DSSP(dsspPath + line[0:4] + \".dssp\")\n",
    "\t\t\tdescription = \"> \" + d.identifier + \"|\" + d.protein + \"|\" + d.organism\n",
    "\t\t\tseq, struct = d.getSequenceStructure(line[4])\n",
    "\t\t\t\n",
    "\t\t\toutFile.writelines(l + \"\\n\" for l in [description,seq,struct])\n",
    "\t\t\n",
    "with open(joinpath(\"dataset\", \"CATH_info_test.txt\")) as infoFile:\n",
    "\twith open(joinpath(\"dataset\", \"CATH_info_test-PARSED.txt\"), 'w') as outFile:\n",
    "\t\tdsspTestPath = joinpath(\"dataset\", \"dssp_test\", \"\")\n",
    "\t\tfor line in infoFile.readlines():\n",
    "\t\t\td = DSSP(dsspTestPath + line[0:4] + \".dssp\")\n",
    "\t\t\tdescription = \"> \" + d.identifier + \"|\" + d.protein + \"|\" + d.organism\n",
    "\t\t\tseq, struct = d.getSequenceStructure(line[4])\n",
    "\t\t\t\n",
    "\t\t\toutFile.writelines(l + \"\\n\" for l in [description,seq,struct])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Training GOR III\n",
    "\n",
    "We can now create our `GOR3` instance and train it with the data :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gor3Pred = GOR3()\n",
    "\n",
    "with open(joinpath(\"dataset\", \"CATH_info-PARSED.txt\")) as inFile:\n",
    "\tindex = 0\n",
    "\tsequence = \"\"\n",
    "\tfor line in inFile.readlines():\n",
    "\t\tline = line.strip().upper()\n",
    "\t\tif not (line==\"\" or line[0]==\">\"):\n",
    "\t\t\t#Line is a sequence\n",
    "\t\t\tif index % 2 == 0:\n",
    "\t\t\t\tsequence = line\n",
    "\t\t\t#Line is a structure\n",
    "\t\t\telse:\n",
    "\t\t\t\tgor3Pred.train(sequence, line)\n",
    "\t\t\tindex += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Predicting structures\n",
    "\n",
    "Now that it's trained, our object can be used to issue structure predictions. Here are the results for a few new sequences (real structure on top, predicted structure at the bottom).\n",
    "Explain Q3, MCC and ROC curves."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------- STRUCTURE (top) vs PREDICTION (bottom) --------\n",
      "True Positive Rate: 0.31\n",
      "\n",
      "CCCTTTCCTTCCCCHHHHHHHHHHHHHHCTTTEEEEEEEECTTCCEEEEEEECCCCCCCCEEEEEECCCTTCHHHHHHHHH\n",
      "    :            :::::::::::         : :      ::::::        :::: :   :     ::: ::\n",
      "ETTETEETEEEEETETTHHHHHHHHHHHHHHHHHHHHETEEEHTETEEEEEETETTTTETEEEETETEHTETEEEHHHTHH\n",
      "\n",
      "HHHHHHHHHTTTCHHHHHHHHHCEEEEECCCCHHHHHHHHHTCTTCCCCCCCCCCCCCCCCCHHHCCCCCTTCCCCECCTT\n",
      ":  :      ::       : :  :::   :            :                           :       : \n",
      "HEEHTEETTETTTTTECEEHTHHTEEETEECTTTTEEECTTHTTHEETEHTEEETTEEETETETTTETHTETTETETTTTE\n",
      "\n",
      "TCTTECCCCTTCCHHHHHHHHHHHHHCCEEEEEEEEECCCEEEECCCCCCCCCTTHHHHHHHHHHHHHHHHHHHCCCCEEE\n",
      "  : :     :                 : :::       ::::         ::: :: ::::::::::::      :: \n",
      "ETTEEETHEETEEEEETTEETEETTETTETEEEHHHHHTTEEEETETEEEETETTHEHHTHHHHHHHHHHHHEETETEEET\n",
      "\n",
      "EEHHHHCCCCCCCHHHHHHHTTCCEEEEEEECCCCCCHHHCCHHHHHHHHHHHHHHHHHHHHHHHHC\n",
      " :         :         :  ::::: :  :        :::::  ::::::  :         \n",
      "TEEEEEEEEEECTEETEEETETETEEEEETETTCTTHTEEETHHHHHCCHHHHHHTEHEEETEEETT\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open(joinpath(\"dataset\", \"CATH_info_test-PARSED.txt\")) as inFile:\n",
    "\tindex = 0\n",
    "\tsequence = \"\"\n",
    "\tfor line in inFile.readlines():\n",
    "\t\tline = line.strip().upper()\n",
    "\t\tif not (line==\"\" or line[0]==\">\"):\n",
    "\t\t\t#Line is a sequence\n",
    "\t\t\tif index % 2 == 0:\n",
    "\t\t\t\tsequence = line\n",
    "\t\t\t#Line is a structure\n",
    "\t\t\telse:\n",
    "\t\t\t\tstructure = line\n",
    "\t\t\t\tprediction = gor3Pred.predict(sequence, structure)\n",
    "\t\t\t\tinter = \"\".join([\":\" if s1==s2 else \" \" for s1,s2 in zip(structure, prediction)])\n",
    "\t\t\t\t\n",
    "\t\t\t\tprint(\"-------- STRUCTURE (top) vs PREDICTION (bottom) --------\")\n",
    "\t\t\t\tprint(\"True Positive Rate:\", round(inter.count(\":\")/len(inter), 2))\n",
    "\t\t\t\tprint()\n",
    "\t\t\t\tchunk = 80\n",
    "\t\t\t\tfor start in range(0, len(structure), chunk):\n",
    "\t\t\t\t\tstop = start+chunk+1 if start+chunk+1<=len(structure) else len(structure)\n",
    "\t\t\t\t\tprint(structure[start:stop])\n",
    "\t\t\t\t\tprint(inter[start:stop])\n",
    "\t\t\t\t\tprint(prediction[start:stop])\n",
    "\t\t\t\t\tprint()\n",
    "\t\t\t\tprint()\n",
    "\t\t\t\t\n",
    "\t\t\tindex += 1\n",
    "\tgor3Pred.plotROC()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking prediction quality\n",
    "\n",
    "let's check it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "q3, mcc = gor3Pred.getQuality()\n",
    "\n",
    "print(\"Q3:\", round(q3, 2))\n",
    "print(\"MCC:\", round(mcc, 2))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
