{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------- Alignment ----------\n",
      "Type       : local\n",
      "Score      : 223\n",
      "Identity   : 48 (55.17%)\n",
      "Similarity : 69 (79.31%)\n",
      "Gaps       : 2\n",
      "\n",
      "Upper Sequence : sp|P78352|65-151\n",
      "Lower Sequence : sp|P78352|160-246\n",
      "\n",
      "EITLERGNSGLGFSIAGGTDNPHI-GDDPSIFITKIIPGGAAAQDGRLRVNDSILFVNEVDVREVTHSAAVEALKEAGSI\n",
      ":: : .: .:::::::::. : :: ::. ::..:::: :::: .::::...:.:: ::.: ...: :. :: :::.. ..\n",
      "EIKLIKGPKGLGFSIAGGVGNQHIPGDN-SIYVTKIIEGGAAHKDGRLQIGDKILAVNSVGLEDVMHEDAVAALKNTYDV\n",
      "\n",
      "\n",
      "VRLYVMR\n",
      ": : : .\n",
      "VYLKVAK\n",
      "\n",
      "---------- Alignment ----------\n",
      "Type       : local-suboptimal\n",
      "Score      : 81\n",
      "Identity   : 2 (20.00%)\n",
      "Similarity : 6 (60.00%)\n",
      "Gaps       : 4\n",
      "\n",
      "Upper Sequence : sp|P78352|65-151\n",
      "Lower Sequence : sp|P78352|160-246\n",
      "\n",
      "IGDD--P--S\n",
      ".:..  :  .\n",
      "VGNQHIPGDN\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sys import path\n",
    "from os import path as ospath\n",
    "path.append(ospath.join(\"..\", \"..\", \"src\"))\n",
    "\n",
    "from sequence import Sequence, loadFasta\n",
    "from aminoAcid import AminoAcid\n",
    "from score import Score\n",
    "from alignMatrix import AlignMatrix\n",
    "from math import log, ceil"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Project 2 - Substitution matrices\n",
    "\n",
    "In the past project we've made extensive use of Soring systems in order to achieve sequence alignment. The scores assigned to each amino acid pair were not understood, we simply used standard ones. The scoring systems we used are called **substitution matrices** as they provide scores for each possible amino acid substitution. When comparing two sequences, one of the following is true:\n",
    "* **Evolutionary model**: sequences B and A do not have a common origin. Therefore B can be seen as completely random when compared to A. This does not mean B has to be completely different than A : amino acids $a$ and $b$ could still align by chance. We can calculate the probability of that happening by looking at the amino acid frequencies $p_a$ and $p_b$ in the population.\n",
    "$$Prob(a, b) = p_a \\cdot p_b $$\n",
    "\n",
    "* **Random model**: sequences B and A have a common origin. Therefore B is not random, and a good alignment between A and B will provide us with the result of each of their mutations. The probability of amino acids $a$ and $b$ being aligned depends on the likelihood of substitutions from $a$ to $b$ or $b$ to $a$.\n",
    "$$Prob(a, b) = q_{a,b}$$\n",
    "\n",
    "The goal of this section is to be able to tell these cases apart. In order to do that, we'll construct substitution matrices, assigning to each amino acid pair a score equal to the **log-odds ratio** $\\frac{q_{a,b}}{p_a \\cdot p_b}$.\n",
    "\n",
    "## BLOCKS\n",
    "\n",
    "`BLOCKS` is a database of amino acid sequences. It was used to construct the standard BLOSUM matrices. We used it to retrieve two families of sequences : the SH3 and PDZ domains. These come in multiple \"blocks\" of gapless aligned sequences of the same size. We quickly formatted these informations into .fasta files compatible with our previous project. With them, we'll be able to construct our own BLOSUM matrices specific to these domains.\n",
    "\n",
    "## BLOSUM\n",
    "\n",
    "`BLOSUM` or BLOcks SUbstitution Matrix uses the well-conserved (gapless and aligned) sequences found in `BLOCKS` to determine the numerical values of $p_a$, $p_b$ and $q_{a,b}$, based on the required identity between them. For example, `BLOSUM 62` is the `BLOSUM` that was calculated for a required identity of 62%. Here's an overview of the algorithm for calculating the log-odds ratios with an identity of $X$% :\n",
    "\n",
    "* Separate all sequences into $t$ groups $G_{i}$, such that all sequences $s_{j} \\in G_{i}$ in each group have at least $X$ percent identity with some other sequence of the same group. We assume the sequences have she same lenght $n$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def belongs(outSequence, group, minMatches):\n",
    "\t\"\"\"\n",
    "\tReturns True if there's at least 'minMatches' matches between outSequence and any sequence \n",
    "\tfrom 'group'; False otherwise.\n",
    "\tSize of all sequences is assumed to be equal.\n",
    "\t\"\"\"\n",
    "\tfor inSequence in group:\n",
    "\t\tmatches = 0\n",
    "\t\tfor i in range(len(outSequence)):\n",
    "\t\t\tif outSequence[i] == inSequence[i]:\n",
    "\t\t\t\tmatches += 1\n",
    "\t\t\tif matches >= minMatches: #As soon as a match is found\n",
    "\t\t\t\treturn True\n",
    "\treturn False\n",
    "\n",
    "\t\n",
    "def makeGroupsFromFasta(path, requiredIdentityPercent):\n",
    "\t\"\"\"\n",
    "\tLoads Sequences from file at 'path' and separate them in groups with an identity of \n",
    "\tat least 'requiredIdentityPercent'.\n",
    "\tReturns a list representing the groups as lists of Sequence objects.\n",
    "\t\"\"\"\n",
    "\tsequences = [seq for seq in loadFasta(path)]\n",
    "\tgroups = [[sequences[0]]] #First sequence is assigned to first group\n",
    "\t\n",
    "\tseqSize = len(sequences[0]) #Size of the sequences\n",
    "\t\n",
    "\t#Number of matches required to achieve requiredIdentityPercent\n",
    "\tminMatches = ceil((requiredIdentityPercent/100)*seqSize)\n",
    "\t\n",
    "\t#For each outSequence not yet in a group\n",
    "\tfor outSequence in sequences[1:]:\n",
    "\t\tgroupFound = False #has a group been found ?\n",
    "\t\tgroupIndex = 0 #index of the group we're looking in\n",
    "\t\t\n",
    "\t\t#Look for a group where outSequence belongs\n",
    "\t\twhile (not groupFound) and groupIndex < len(groups):\n",
    "\t\t\tif belongs(outSequence, groups[groupIndex], minMatches):\n",
    "\t\t\t\tgroups[groupIndex].append(outSequence)\n",
    "\t\t\t\tgroupFound = True\n",
    "\t\t\telse:\n",
    "\t\t\t\tgroupIndex += 1 #Move on to next group\n",
    "\t\t\n",
    "\t\t#If no group works, create a new one\n",
    "\t\tif not groupFound:\n",
    "\t\t\tgroups.append([outSequence])\n",
    "\t\n",
    "\treturn groups\n",
    "\n",
    "def valueDictsFromGroups(groups):\n",
    "\t\"\"\"\n",
    "\tTransforms each group from 'groups' into a list of dictionaries, one for each Sequence column.\n",
    "\tThe dictionaries map each AminoAcid found in that column to their count.\n",
    "\tReturns the list of dictionaries and a list of the size (in Sequences) of their groups.\n",
    "\t\"\"\"\n",
    "\tseqSize = len(groups[0][0])\n",
    "\tgroupCount = len(groups)\n",
    "\t\n",
    "\tgroupValues = []\n",
    "\tgroupSizes = []\n",
    "\t\n",
    "\tfor group in groups: #For each group\n",
    "\t\tgroupAAs = []\n",
    "\t\tgroupSize = len(group) #Size of group (nÂ°sequences)\n",
    "\t\t\n",
    "\t\tfor col in range(seqSize): #For each column\n",
    "\t\t\tgroupCol = {}\n",
    "\t\t\t\n",
    "\t\t\tfor seq in group: #For each sequence in group\n",
    "\t\t\t\ttry:\n",
    "\t\t\t\t\tgroupCol[seq[col]] += 1 #Increment count\n",
    "\t\t\t\texcept:\n",
    "\t\t\t\t\tgroupCol[seq[col]] = 1\n",
    "\t\t\t\n",
    "\t\t\tgroupAAs.append(groupCol)\n",
    "\t\tgroupValues.append(groupAAs)\n",
    "\t\tgroupSizes.append(groupSize)\n",
    "\treturn groupValues, groupSizes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Calculate the weighted frequencies of all amino acid pairs, where $Count(G_{i_k}, a)$ is the number of occurrences of amino acid $a$ in column $k$ of sequences from $G_i$, $Size(G_i)$ is the number of sequences in group $G_i$ :\n",
    "\n",
    "$$f_{a,b} = \\sum_{k=1}^{n}{\\sum_{i=1}^{t-1}{\\sum_{j=i+1}^{t}{ \\frac{Count(G_{i_k},a)}{Size(G_i)} \\cdot \\frac{Count(G_{j_k},b)}{Size(G_j)} + \\frac{Count(G_{i_k},b)}{Size(G_i)} \\cdot \\frac{Count(G_{j_k},a)}{Size(G_j)}}}}$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def getFrequencies(groupValues, groupSizes):\n",
    "\t\"\"\"\n",
    "\tEvaluates the frequencies of AminoAcids within columns of groups in 'groupValues'.\n",
    "\tFrequencies are weighted according to group sizes in 'groupSizes'.\n",
    "\tReturns two dictionaries and a number:\n",
    "\t\t-'freqPairs' maps pairs of AminoAcids to their frequencies\n",
    "\t\t-'freqSingle' maps single AminoAcids to their frequencies\n",
    "\t\t-'freqSum' is the sum of all frequencies\n",
    "\t\"\"\"\n",
    "\tseqSize = len(groupValues[0]) #Size of the Sequences\n",
    "\tgroupCount = len(groupSizes) #Number of groups\n",
    "\t\n",
    "\tfreqPairs = {} #frequencies of amino acid pairs (fAB)\n",
    "\t\n",
    "\t#Frequencies of single amino acids (fA)\n",
    "\tfreqSingle = {AminoAcid(aa):0 for aa in AminoAcid.getAllNames()} \n",
    "\t\n",
    "\tfreqSum = 0 #Sum of frequencies  sum(fAB)\n",
    "\t\n",
    "\tfor col in range(seqSize): #Each column\n",
    "\t\tfor groupAIndex in range(groupCount-1): #Each groupA\n",
    "\t\t\tgroupA = groupValues[groupAIndex]\n",
    "\t\t\tgroupASize = groupSizes[groupAIndex]\n",
    "\t\t\t\n",
    "\t\t\tfor groupBIndex in range(groupAIndex+1, groupCount): #Each further groupB\n",
    "\t\t\t\tgroupB = groupValues[groupBIndex]\n",
    "\t\t\t\tgroupBSize = groupSizes[groupBIndex]\n",
    "\t\t\t\t\n",
    "\t\t\t\tfor aaA, aaACount in groupA[col].items(): #Each AA from groupA\n",
    "\t\t\t\t\taaAFreq = aaACount / groupASize #Its frequency within groupA\n",
    "\t\t\t\t\t\n",
    "\t\t\t\t\tfor aaB, aaBCount in groupB[col].items(): #Each AA from groupB\n",
    "\t\t\t\t\t\taaBFreq = aaBCount / groupBSize #Its frequency within groupB\n",
    "\t\t\t\t\t\t\n",
    "\t\t\t\t\t\taaPairFreq = aaAFreq * aaBFreq #Pair frequency\n",
    "\t\t\t\t\t\tfreqSum += aaPairFreq\t#Sum of all frequencies\t\t\t\n",
    "\t\t\t\t\t\tfreqSingle[aaA] += aaPairFreq/2\n",
    "\t\t\t\t\t\tfreqSingle[aaB] += aaPairFreq/2\n",
    "\t\t\t\t\t\t\n",
    "\t\t\t\t\t\t#Index is unique to this pair\n",
    "\t\t\t\t\t\tpairIndex = (aaA, aaB) if aaA > aaB else (aaB, aaA)\n",
    "\t\t\t\t\t\ttry:\n",
    "\t\t\t\t\t\t\tfreqPairs[pairIndex] += aaPairFreq\n",
    "\t\t\t\t\t\texcept:\n",
    "\t\t\t\t\t\t\tfreqPairs[pairIndex] = aaPairFreq\n",
    "\t\n",
    "\treturn freqPairs, freqSingle, freqSum"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Calculate the probability of occurrences $a, b$ in the evolutionary model as the frequence of pairs $(a,b)$ divided by the frequence of all possible pairs\n",
    "$$q_{a,b} = \\frac{f_{a,b}}{\\sum_{1\\leq a \\leq b}{f_{a,b}}}$$\n",
    "\n",
    "* Calculate the probability of occurrences $a, b$ in the random model, based on the observed frequencies of each single amino acid\n",
    "$$\n",
    "e_{a,b} =\n",
    "\\left\\{\n",
    "    \\begin{array}{ll}\n",
    "      p_a^2 \\quad \\textrm{if} \\quad a = b\\\\\n",
    "      2 p_a p_b \\quad \\textrm{if} \\quad a \\neq b\n",
    "    \\end{array}\n",
    "\\right.\n",
    "\\quad\n",
    "\\textrm{where}\n",
    "\\quad\n",
    "p_a = q_{a,a} + \\frac{1}{2}\\sum_{a \\neq b}{q_{a,b}}\n",
    "$$\n",
    "\n",
    "* Compute the log-odds ratio of all probabilities for the evolutionary and random models. This gives us a negative value (a bad score) if the random model is more likely, and a positive value the other way around. This value is the one used in BLOSUM (except we round it to the nearest integer).\n",
    "$$s_{a,b} = 2 \\; \\log_{2}{\\frac{q_{a,b}}{e_{a,b}}}$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def sumFrequenciesToProb(freqPairsList, freqSingleList, freqSumList):\n",
    "\t\"\"\"\n",
    "\tSums all frequencies in provided lists, and transforms them to probabilities according\n",
    "\tto the sum of frequencies found in 'freqSumList'.\n",
    "\t\"\"\"\n",
    "\tfSum = sum(freqSumList) #Absolute sum of all frequencies\n",
    "\tprobPairs, probSingle = {}, {} #Probabilities for pairs and single AAs\n",
    "\t\n",
    "\tfor freqPairs, freqSingle in zip(freqPairsList, freqSingleList):\n",
    "\t\tfor key,value in freqPairs.items():\n",
    "\t\t\t#Sum all frequencies for matching AA pairs, divided by fSum\n",
    "\t\t\ttry:\n",
    "\t\t\t\tprobPairs[key] += value / fSum \n",
    "\t\t\texcept:\n",
    "\t\t\t\tprobPairs[key] = value / fSum\n",
    "\t\tfor key,value in freqSingle.items():\n",
    "\t\t\t#Sum all frequencies for matching AAs, divided by fSum\n",
    "\t\t\ttry:\n",
    "\t\t\t\tprobSingle[key] += value / fSum\n",
    "\t\t\texcept:\n",
    "\t\t\t\tprobSingle[key] = value / fSum\n",
    "\t\n",
    "\treturn probPairs, probSingle\n",
    "\t\t\n",
    "\t\n",
    "def blosumFromProbabilities(probPairs, probSingle, requiredIdentityPercent):\n",
    "\t\"\"\"\n",
    "\tFills and returns a Score according to the BLOSUM algorithm, from the probabilities\n",
    "\tof AA pairs and singletons provided in 'probPairs' and 'probSingle'.\n",
    "\t\"\"\"\n",
    "\t#Create empty Score, ignoring AAs B,Z,J,U,O\n",
    "\tscoreMatrix = Score(\"\", \"BLOSUM{}\".format(requiredIdentityPercent), \"BZJUO\")\n",
    "\t\n",
    "\tfor key, qAB in probPairs.items():\n",
    "\t\t#qAB is the evolutionary probability of the AA pair (A,B)\n",
    "\t\taaA, aaB = key #Both amino acids\n",
    "\t\tpA, pB = probSingle[aaA], probSingle[aaB] #Their single probabilities\n",
    "\t\t\n",
    "\t\t#eAB is the random probability of the AA pair (A, B) given their single probabilities\n",
    "\t\teAB = pA * pA if aaA == aaB else 2 * pA * pB\n",
    "\t\t\n",
    "\t\t#The BLOSUM score for this pair is the log-odds-ratio of evolutionary and random prob.\n",
    "\t\tsAB = int(round(2 * log(qAB / eAB, 2)))\n",
    "\t\tscoreMatrix.setScore(aaA, aaB, sAB) #Fill the matrix\n",
    "\t\n",
    "\treturn scoreMatrix\n",
    "\t\t\t\n",
    "\t\t\n",
    "\t\t\t\n",
    "def blosumFromFasta(requiredIdentityPercent, *filepaths):\n",
    "\t\"\"\"\n",
    "\tCreates and returns a Score for all sequences in the provided 'filepaths',\n",
    "\tusing the BLOSUM approach with an identity of at least 'requiredIdentityPercent'.\n",
    "\tEach file is grouped independently and only then their weighted probabilities are merged.\n",
    "\t\"\"\"\n",
    "\t#Results for each different files are first stored in lists\n",
    "\tfreqPairsList, freqSingleList, freqSumList = [], [], [] \n",
    "\t\n",
    "\tfor path in filepaths: #for each .fasta file\n",
    "\t\tgroups = makeGroupsFromFasta(path, requiredIdentityPercent) #groups\n",
    "\t\tgroupValues, groupSizes = valueDictsFromGroups(groups) #groups as dicts\n",
    "\t\tfreqPairs, freqSingle, freqSum = getFrequencies(groupValues, groupSizes) #frequencies\n",
    "\t\t\n",
    "\t\tfreqPairsList.append(freqPairs) #Append results\n",
    "\t\tfreqSingleList.append(freqSingle)\n",
    "\t\tfreqSumList.append(freqSum)\n",
    "\t\n",
    "\t#Merge (sum) results together\n",
    "\tprobPairs, probSingle = sumFrequenciesToProb(freqPairsList, freqSingleList, freqSumList)\n",
    "\t#Create the BLOSUM matrix\n",
    "\tblosum = blosumFromProbabilities(probPairs, probSingle, requiredIdentityPercent)\n",
    "\treturn blosum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SH3 Family\n",
      "---------- BLOSUM40 ----------\n",
      "A   2   \n",
      "C   3   6   \n",
      "D   -1  -10 2   \n",
      "E   0   -8  2   3   \n",
      "F   -1  3   -4  -4  4   \n",
      "G   0   0   3   -1  -3  4   \n",
      "H   0   -7  -2  0   -5  -3  2   \n",
      "I   -2  0   -2  -2  2   -2  0   1   \n",
      "K   1   -8  0   2   -3  -3  1   -1  2   \n",
      "L   -2  -3  -3  -2  0   -2  0   2   -1  2   \n",
      "M   -1  -1  -1  -2  2   -5  -1  2   -3  2   0   \n",
      "N   1   1   1   0   2   1   -2  -3  2   -4  -8  3   \n",
      "P   0   2   0   0   -5  -1  -2  -3  0   -3  -3  0   5   \n",
      "Q   -1  -8  1   1   -3  1   0   -1  1   -2  -4  0   0   0   \n",
      "R   -1  -4  0   1   -1  -2  1   0   0   -2  -1  0   0   0   1   \n",
      "S   1   -3  0   -1  -1  0   2   -2  0   -2  -2  2   1   -1  2   1   \n",
      "T   2   2   -2  -1  -2  -1  1   0   0   1   1   -1  0   0   0   0   1   \n",
      "V   -1  1   -1  -2  1   -1  -2  1   -2  2   1   -8  0   -2  0   -1  1   1   \n",
      "W   -1  -1  -1  -8  1   -12 4   1   -3  1   -6  -10 -16 1   -2  -2  0   -1  7   \n",
      "Y   -2  1   -2  -8  3   -2  1   -1  0   0   -3  1   1   1   1   1   -2  0   3   2   \n",
      "X   1   -6  -8  3   -8  -10 -1  -4  2   -4  2   1   5   -5  0   0   -6  -5  -6  -6  8   \n",
      "\n",
      "    A   C   D   E   F   G   H   I   K   L   M   N   P   Q   R   S   T   V   W   Y   X   \n",
      "---------- BLOSUM70 ----------\n",
      "A   4   \n",
      "C   1   3   \n",
      "D   -1  -2  4   \n",
      "E   -1  -1  2   3   \n",
      "F   -3  1   -5  -4  5   \n",
      "G   0   -2  0   -2  -7  6   \n",
      "H   -1  1   0   0   0   -2  2   \n",
      "I   -3  -1  -5  -2  1   -6  -1  3   \n",
      "K   0   0   -1  1   -3  -3  1   -1  3   \n",
      "L   -3  0   -4  -3  1   -6  -1  2   -2  4   \n",
      "M   -1  0   -2  -1  0   -4  0   1   0   2   2   \n",
      "N   0   -1  1   0   -3  1   0   -3  0   -3  -2  4   \n",
      "P   0   -3  -1  -2  -6  -2  -2  -4  -1  -4  -3  -1  7   \n",
      "Q   0   -1  1   2   -3  -2  1   -1  2   -2  0   0   -2  2   \n",
      "R   -1  0   -1  1   -2  -3  1   -1  2   -2  0   0   -2  2   2   \n",
      "S   1   -1  1   0   -4  -1  0   -3  0   -3  -2  2   1   0   1   4   \n",
      "T   0   0   0   1   -3  -3  1   0   1   -1  0   1   -1  1   1   1   2   \n",
      "V   0   1   -4  -2  0   -4  -1  2   -1  1   0   -2  -4  -1  0   -2  0   3   \n",
      "W   -4  1   -5  -5  0   -10 -1  -2  -5  0   0   -5  -8  -4  -4  -4  -3  -3  10  \n",
      "Y   -2  3   -4  -3  3   -5  3   -1  -1  -1  0   -2  -5  -2  -1  -2  -1  -2  2   5   \n",
      "X   -1  0   -1  1   -3  -3  -1  1   0   1   1   -1  1   1   0   -1  0   1   -6  -1  7   \n",
      "\n",
      "    A   C   D   E   F   G   H   I   K   L   M   N   P   Q   R   S   T   V   W   Y   X   \n",
      "PDZ Family\n",
      "---------- BLOSUM40 ----------\n",
      "A   0   \n",
      "C   1   -2  \n",
      "D   -1  -2  5   \n",
      "E   0   0   2   0   \n",
      "F   0   1   -2  1   2   \n",
      "G   0   0   -4  -3  -2  3   \n",
      "H   0   -2  -1  0   -7  0   0   \n",
      "I   0   0   -2  0   0   -3  0   2   \n",
      "K   0   0   -1  0   0   -2  1   0   0   \n",
      "L   1   1   -2  1   1   -2  0   1   1   1   \n",
      "M   1   -2  1   0   -2  -2  2   0   1   0   2   \n",
      "N   -1  -1  1   -2  -2  -1  1   -1  1   -2  -1  4   \n",
      "P   0   0   0   1   1   -2  -1  1   -1  1   2   -2  1   \n",
      "Q   -1  -3  1   1   1   -3  1   2   0   1   1   -1  1   1   \n",
      "R   0   1   -1  0   2   -3  -2  1   1   1   0   0   1   0   2   \n",
      "S   1   1   -2  -1  1   0   -1  0   0   0   0   0   0   -1  0   0   \n",
      "T   1   1   0   2   0   -2  1   1   0   1   0   -3  2   -1  1   0   2   \n",
      "V   0   0   -1  1   1   -2  0   2   1   1   0   -1  1   2   1   -1  0   1   \n",
      "W   0   2   -2  -3  -1  0   1   2   0   0   1   0   -3  0   0   1   -5  1   3   \n",
      "Y   0   -1  2   2   0   -3  1   -1  0   0   3   -1  2   1   0   -1  1   0   0   -1  \n",
      "X   0   -1  -1  1   2   -3  -1  0   0   0   -7  0   1   2   1   -1  0   2   -15 2   0   \n",
      "\n",
      "    A   C   D   E   F   G   H   I   K   L   M   N   P   Q   R   S   T   V   W   Y   X   \n",
      "---------- BLOSUM70 ----------\n",
      "A   2   \n",
      "C   0   1   \n",
      "D   -2  -3  5   \n",
      "E   1   -1  0   2   \n",
      "F   1   1   -3  -1  2   \n",
      "G   -1  1   -2  -2  -1  3   \n",
      "H   1   -1  0   1   0   -1  1   \n",
      "I   -1  0   -6  -2  1   -7  -1  3   \n",
      "K   1   -1  -2  2   -1  -1  1   -2  2   \n",
      "L   0   0   -4  -1  0   -3  -1  2   -1  2   \n",
      "M   0   0   -2  0   1   -3  0   1   0   1   1   \n",
      "N   -1  0   2   -1  -3  1   0   -6  -1  -4  -3  5   \n",
      "P   1   0   -1  0   1   -3  1   -2  1   0   1   0   4   \n",
      "Q   0   -1  -1  2   -1  -2  1   -1  2   -1  1   0   0   2   \n",
      "R   1   -1  -2  2   0   -2  1   -1  2   0   1   -1  1   2   2   \n",
      "S   1   0   -2  1   0   0   0   -3  1   -1  0   -1  2   0   0   2   \n",
      "T   0   0   -2  1   0   -3  1   0   1   1   1   -2  1   1   1   0   2   \n",
      "V   0   0   -5  0   1   -5  -1  2   -1  2   1   -5  -1  0   0   -2  1   2   \n",
      "W   0   1   -5  0   0   0   0   1   0   0   0   -3  -1  0   1   0   0   1   2   \n",
      "Y   0   0   -1  1   1   -2  1   -1  1   0   1   -2  2   1   1   0   1   0   1   2   \n",
      "X   -1  -1  -1  1   0   -1  1   -1  1   -1  0   0   2   1   1   0   2   1   -2  1   0   \n",
      "\n",
      "    A   C   D   E   F   G   H   I   K   L   M   N   P   Q   R   S   T   V   W   Y   X   \n"
     ]
    }
   ],
   "source": [
    "sh3Paths = [\"SH3-A.fasta\", \"SH3-B.fasta\", \"SH3-C.fasta\", \"SH3-D.fasta\"]\n",
    "print(\"SH3 Family\")\n",
    "blosum40sh3 = blosumFromFasta(40, *sh3Paths)\n",
    "print(blosum40sh3)\n",
    "blosum70sh3 = blosumFromFasta(70, *sh3Paths)\n",
    "print(blosum70sh3)\n",
    "\n",
    "print(\"PDZ Family\")\n",
    "pdzPaths = [\"PDZ-A.fasta\", \"PDZ-B.fasta\"]\n",
    "blosum40pdz = blosumFromFasta(40, *pdzPaths)\n",
    "print(blosum40pdz)\n",
    "blosum70pdz = blosumFromFasta(70, *pdzPaths)\n",
    "print(blosum70pdz)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a reminder, here is BLOSUM 62 :\n",
    "```\n",
    "A 4 \n",
    "R -1   5 \n",
    "N -2   0   6 \n",
    "D -2  -2   1   6 \n",
    "C  0  -3  -3  -3   9 \n",
    "Q -1   1   0   0  -3   5 \n",
    "E -1   0   0   2  -4   2   5 \n",
    "G  0  -2   0  -1  -3  -2  -2   6 \n",
    "H -2   0   1  -1  -3   0   0  -2   8 \n",
    "I -1  -3  -3  -3  -1  -3  -3  -4  -3   4 \n",
    "L -1  -2  -3  -4  -1  -2  -3  -4  -3   2   4 \n",
    "K -1   2   0  -1  -3   1   1  -2  -1  -3  -2   5 \n",
    "M -1  -1  -2  -3  -1   0  -2  -3  -2   1   2  -1   5 \n",
    "F -2  -3  -3  -3  -2  -3  -3  -3  -1   0   0  -3   0   6 \n",
    "P -1  -2  -2  -1  -3  -1  -1  -2  -2  -3  -3  -1  -2  -4   7 \n",
    "S  1  -1   1   0  -1   0   0   0  -1  -2  -2   0  -1  -2  -1   4 \n",
    "T  0  -1   0  -1  -1  -1  -1  -2  -2  -1  -1  -1  -1  -2  -1   1   5 \n",
    "W -3  -3  -4  -4  -2  -2  -3  -2  -2  -3  -2  -3  -1   1  -4  -3  -2  11 \n",
    "Y -2  -2  -2  -3  -2  -1  -2  -3   2  -1  -1  -2  -1   3  -3  -2  -2   2   7 \n",
    "V  0  -3  -3  -3  -1  -2  -2  -3  -3   3   1  -2   1  -1  -2  -2   0  -3  -1   4 \n",
    "B -2  -1   3   4  -3   0   1  -1   0  -3  -4   0  -3  -3  -2   0  -1  -4  -3  -3   4\n",
    "Z -1   0   0   1  -3   3   4  -2   0  -3  -3   1  -1  -3  -1   0  -1  -3  -2  -2   1   4\n",
    "X  0  -1  -1  -1  -2  -1  -1  -1  -1  -1  -1  -1  -1  -1  -2   0   0  -2  -1  -1  -1  -1  -1\n",
    "   A   R   N   D   C   Q   E   G   H   I   L   K   M   F   P   S   T   W   Y   V   B   Z   X\n",
    "```\n",
    "\n",
    "Values are very different, which is normal since this matrix was built for a small, specific subset of the BLOCKS database and with a different required identity. Diagonal values also differ significantly. New substitutions are allowed (like between P and A), other substitutions no longer are (like all substitutions to and from Z, which is not present in the SH3 or PDZ domains).\n",
    "\n",
    "Let's try to do some alignments with these BLOSUM matrices :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Could not find amino acid name ",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32mC:\\Users\\mytra\\Documents\\GitHub\\BioInfo\\src\\aminoAcid.py\u001b[0m in \u001b[0;36m__getIdByName\u001b[0;34m(name)\u001b[0m\n\u001b[1;32m     65\u001b[0m                 \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m---> 66\u001b[0;31m                         \u001b[1;32mreturn\u001b[0m \u001b[0mAminoAcid\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_nameDict\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;31m#get index of name mode\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     67\u001b[0m                 \u001b[1;32mexcept\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: ''",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-339ea46fdb47>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mblosum62\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mScore\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"blosum62.iij\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"BLOSUM 62\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0ma\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mAlignMatrix\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mblosum62\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mb\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mAlignMatrix\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mblosum40pdz\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mc\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mAlignMatrix\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mblosum70pdz\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0msequences\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mseq\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mseq\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mloadFasta\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"PDZ-sequences.fasta\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\mytra\\Documents\\GitHub\\BioInfo\\src\\score.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, path, description, ignore)\u001b[0m\n\u001b[1;32m     14\u001b[0m \t\t\"\"\"\n\u001b[1;32m     15\u001b[0m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_description\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdescription\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_ignore\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mSequence\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mignore\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_matrix\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_aaOrder\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m{\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\mytra\\Documents\\GitHub\\BioInfo\\src\\sequence.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, aminoAcids, description)\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m                 \u001b[1;31m#Format aminoAcids into a list of AminoAcid objects.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_aaList\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__formatAAList\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maminoAcids\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m#List of amino acids\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\mytra\\Documents\\GitHub\\BioInfo\\src\\sequence.py\u001b[0m in \u001b[0;36m__formatAAList\u001b[0;34m(aminoAcids)\u001b[0m\n\u001b[1;32m     44\u001b[0m                                 \u001b[1;32mreturn\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mAminoAcid\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maa\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0maa\u001b[0m \u001b[1;32min\u001b[0m \u001b[0maminoAcids\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m                         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;31m#A single Amino Acid with any name mode\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m---> 46\u001b[0;31m                                 \u001b[1;32mreturn\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mAminoAcid\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maminoAcids\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     47\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m                 \u001b[1;31m#An AminoAcid is copied (from name) and put within a list\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\mytra\\Documents\\GitHub\\BioInfo\\src\\aminoAcid.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, aminoAcid)\u001b[0m\n\u001b[1;32m     54\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m                 \u001b[1;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maminoAcid\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m---> 56\u001b[0;31m                         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_id\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__getIdByName\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maminoAcid\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m#id found from aminoAcid name\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     57\u001b[0m                 \u001b[1;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maminoAcid\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mAminoAcid\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m                         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_id\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0maminoAcid\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_id\u001b[0m \u001b[1;31m#copy of id\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\mytra\\Documents\\GitHub\\BioInfo\\src\\aminoAcid.py\u001b[0m in \u001b[0;36m__getIdByName\u001b[0;34m(name)\u001b[0m\n\u001b[1;32m     66\u001b[0m                         \u001b[1;32mreturn\u001b[0m \u001b[0mAminoAcid\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_nameDict\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;31m#get index of name mode\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     67\u001b[0m                 \u001b[1;32mexcept\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m---> 68\u001b[0;31m                         \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Could not find amino acid name {}\"\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     69\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     70\u001b[0m         \u001b[1;33m@\u001b[0m\u001b[0mstaticmethod\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Could not find amino acid name "
     ]
    }
   ],
   "source": [
    "blosum62 = Score(\"blosum62.iij\", \"BLOSUM 62\")\n",
    "a = AlignMatrix(blosum62)\n",
    "b = AlignMatrix(blosum40pdz)\n",
    "c = AlignMatrix(blosum70pdz)\n",
    "sequences = [seq for seq in loadFasta(\"PDZ-sequences.fasta\")]\n",
    "\n",
    "for alignM in (a,b,c):\n",
    "    for align in a.globalAlign(sequences[0], sequences[1], -12, -2, False):\n",
    "        align.condensed = True\n",
    "        print(align)\n",
    "        break"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
