{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 391,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from copy import deepcopy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1 - Sequence alignment\n",
    "\n",
    "Genetic material such as DNA and proteins comes in sequences made up of different blocks. The study of these sequences and their correlation is invaluable to the field of biology. Since genetic material is likely to mutate over time, it is not easy to be certain of these correlations in most cases. In order to know if two or more sequences have a common origin, we must examine their similarity in a formal way, and be able to quantify it. Since the modifications of the sequences can introduce new blocks as well as delete some, the first step to comparing sequences is aligning them by making the similar regions match. Even then, high levels of similarity do not necessarily imply a common ancestor exists (homology). Our goal in this section is to create a tool that allows us to perform such alignments and give a numeric value of the chance of homology.\n",
    "\n",
    "\n",
    "## ADTs\n",
    "\n",
    "We will first look at an implementation of Abstract Data Types (ADTs) that will allow us to represent Sequences. \n",
    "\n",
    "### AminoAcid\n",
    "\n",
    "The choice of genetic material is proteins. In the case of proteins, the blocks that make up the sequences are called amino acids. Therefore, they are the first thing we must be able to represent. The following tuple lists all amino acids and miscellaneous options."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 392,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#All the names of amino acids, plus gaps and some (partially) undetermined blocks\n",
    "nameGroups = (\n",
    "\t\t(\"none/gap\", \"gap\", \"-\"),\n",
    "\t\t(\"alanine\", \"ala\", \"A\"),\n",
    "\t\t(\"cysteine\", \"cys\", \"C\"),\n",
    "\t\t(\"aspartic acid\", \"asp\", \"D\"),\n",
    "\t\t(\"glutamic acid\", \"glu\", \"E\"),\n",
    "\t\t(\"phenylalanine\", \"phe\", \"F\"),\n",
    "\t\t(\"glycine\", \"gly\", \"G\"),\n",
    "\t\t(\"histidine\", \"his\", \"H\"),\n",
    "\t\t(\"isoleucine\", \"ile\", \"I\"),\n",
    "\t\t(\"lysine\", \"lys\", \"K\"),\n",
    "\t\t(\"leucine\", \"leu\", \"L\"),\n",
    "\t\t(\"methionine\", \"met\", \"M\"),\n",
    "\t\t(\"asparagine\", \"asn\", \"N\"),\n",
    "\t\t(\"proline\", \"pro\", \"P\"),\n",
    "\t\t(\"glutamine\", \"gln\", \"Q\"),\n",
    "\t\t(\"arginine\", \"arg\", \"R\"),\n",
    "\t\t(\"serine\", \"ser\", \"S\"),\n",
    "\t\t(\"threonine\", \"thr\", \"T\"),\n",
    "\t\t(\"valine\", \"val\", \"V\"),\n",
    "\t\t(\"tryptophan\", \"trp\", \"W\"),\n",
    "\t\t(\"tyrosine\", \"tyr\", \"Y\"),\n",
    "\t\t(\"asparagine/aspartic acid\", \"asx\", \"B\"),\n",
    "\t\t(\"glutamine/glutamic acid\", \"glx\", \"Z\"),\n",
    "\t\t(\"leucine/isoleucine\", \"xle\", \"J\"),\n",
    "\t\t(\"selenocysteine\", \"sec\", \"U\"),\n",
    "\t\t(\"pyrrolysine\", \"pyl\", \"O\"),\n",
    "\t\t(\"undetermined\", \"und\", \"X\")\n",
    "\t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following class allows us to create AminoAcid objects that represent any of the amino acids from the list. We can then compare, print, hash, test them with any of the defined methods. Do note that we can create them by copying other AminoAcids or by providing their name in any form (long for complete, medium for 3 chars or short for 1 char)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 393,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "class AminoAcid:\n",
    "\t\"\"\"\n",
    "\tRepresents one of the amino acids that can be found in genetic sequences.\n",
    "\tCan be one of the twenty-two amino acids, four undetermined combinations of possible amino acids, and gaps.\n",
    "\t\"\"\"\n",
    "\t#Tuple of tuples listing amino acid names\n",
    "\t_nameGroups = deepcopy(nameGroups)\n",
    "\t\n",
    "\t#Dictionary mapping name to id\n",
    "\t_nameDict = {nameGroups[id][i]:id for i in range(3) for id in range(len(nameGroups))}\n",
    "\t\n",
    "\t_nameModes = {\"long\":0, \"medium\":1, \"short\":2} #choices for name lenght\n",
    "\t_defaultNameMode = \"short\" #short name by default\n",
    "\t\n",
    "\t\n",
    "\tdef __init__(self, aminoAcid):\n",
    "\t\t\"\"\"\n",
    "\t\tCreates an AminoAcid object representing one of the possible Amino Acids.\n",
    "\t\taminoAcid can be the name of an amino acid, or an AminoAcid object (in which case a copy is created).\n",
    "\t\t\"\"\"\n",
    "\t\tself._id = None\t#id of the amino acid within the name group\n",
    "\t\t\n",
    "\t\tif isinstance(aminoAcid, str):\n",
    "\t\t\tself._id = self.__getIdByName(aminoAcid) #id found from aminoAcid name\n",
    "\t\telif isinstance(aminoAcid, AminoAcid):\n",
    "\t\t\tself._id = aminoAcid._id #copy of id\n",
    "\t\telse:\t\t\t\n",
    "\t\t\traise TypeError(\"aminoAcid must be a string or an AminoAcid object\")\n",
    "\t\t\n",
    "\t\n",
    "\t@staticmethod\n",
    "\tdef __getIdByName(name):\n",
    "\t\ttry:\n",
    "\t\t\treturn AminoAcid._nameDict[name] #get index of name mode\n",
    "\t\texcept:\n",
    "\t\t\traise ValueError(\"Could not find amino acid name {}\".format(name))\n",
    "\t\t\t\n",
    "\t@staticmethod\n",
    "\tdef getAllNames(nameMode=_defaultNameMode):\n",
    "\t\ttry:\n",
    "\t\t\tnameIndex = AminoAcid._nameModes[nameMode] #get index of name mode\n",
    "\t\texcept:\n",
    "\t\t\traise TypeError(\"nameMode must be 'short', 'medium' or 'long'\")\n",
    "\t\t\t\n",
    "\t\tfor aa in AminoAcid._nameGroups[1:]: #we exclude the gap (first item)\n",
    "\t\t\tyield aa[nameIndex]\n",
    "\t\n",
    "\t#Representation\n",
    "\tdef __repr__(self):\n",
    "\t\tnameIndex = AminoAcid._nameModes[self._defaultNameMode]\n",
    "\t\treturn self._nameGroups[self._id][nameIndex] #default name mode\n",
    "\t\t\n",
    "\tdef __str__(self):\n",
    "\t\treturn self.getName() #default name mode\n",
    "\t\n",
    "\tdef getName(self, nameMode=_defaultNameMode):\n",
    "\t\ttry:\n",
    "\t\t\tnameIndex = AminoAcid._nameModes[nameMode] #get index of name mode\n",
    "\t\texcept:\n",
    "\t\t\traise TypeError(\"nameMode must be 'short', 'medium' or 'long'\")\n",
    "\t\t\n",
    "\t\treturn self._nameGroups[self._id][nameIndex]\n",
    "\t\n",
    "\t#Comparison\t\t\n",
    "\tdef __eq__(self, other):\n",
    "\t\treturn self._id == other._id\n",
    "\t\n",
    "\tdef __ne__(self, other):\n",
    "\t\treturn self._id != other._id\n",
    "\t\t\n",
    "\tdef __gt__(self, other):\n",
    "\t\treturn self._id > other._id\n",
    "\t\t\n",
    "\tdef __ge__(self, other):\n",
    "\t\treturn self._id >= other._id\n",
    "\t\t\n",
    "\tdef __lt__(self, other):\n",
    "\t\treturn self._id < other._id\n",
    "\t\n",
    "\tdef __le__(self, other):\n",
    "\t\treturn self._id <= other._id\n",
    "\t\t\n",
    "\tdef isGap(self):\n",
    "\t\treturn self._id == 0\n",
    "\t\n",
    "\t\n",
    "\t#Hashing\n",
    "\tdef __hash__(self):\n",
    "\t\treturn hash(self._id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's test it a little bit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 394,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A\n",
      "True\n",
      "lysine\n"
     ]
    }
   ],
   "source": [
    "a1 = AminoAcid(\"A\")\n",
    "print(a1)\n",
    "a2 = AminoAcid(a1)\n",
    "print(a1 == a2)\n",
    "a3 = AminoAcid(\"K\")\n",
    "print(a3.getName(\"long\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequence\n",
    "\n",
    "That's good and all, but we don't want to align single amino acids do we ? What we want to align is sequences. We can view a Sequence as an ordered list of AminoAcid objects. It should be iterable : we'll want to go through, access, and count its items. We'll also want to change them, insert new ones, delete some, and do all that transparently, as we would with a list. Finally it would be useful to check whether a sequence contains some other sub-sequence or single amino acid. The following is a class that does just that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "class Sequence:\n",
    "\t\"\"\"\n",
    "\tRepresents a sequence of amino acids.\n",
    "\t\"\"\"\n",
    "\t\n",
    "\tdef __init__(self, aminoAcids=None, description=\"\"):\n",
    "\t\t\"\"\"\n",
    "\t\tCreates a Sequence object that represents the amino acid sequence contained in aminoAcids.\n",
    "\t\taminoAcids can be one of the following :\n",
    "\t\t- None, meaning the Sequence is empty (default)\n",
    "\t\t- an AminoAcid object\n",
    "\t\t- a string of X AminoAcid short (uppercase) names or 1 AminoAcid name\n",
    "\t\t- a list containing AminoAcid objects and/or strings of individual AminoAcid names\n",
    "\t\t\"\"\"\n",
    "\t\t\n",
    "\t\tself._nameMode = \"short\" #the way in which AA names are displayed\n",
    "\t\tself._separator = \"\" #how to separate AA names when displayed\n",
    "\t\tself._description = description #description of the sequence\n",
    "\t\tif self._description == \"\" and isinstance(aminoAcids, Sequence):\n",
    "\t\t\tself._description = aminoAcids.getDescription()\n",
    "\t\t#Format aminoAcids into a list of AminoAcid objects.\n",
    "\t\tself._aaList = self.__formatAAList(aminoAcids) #List of amino acids\n",
    "\t\n",
    "\t\n",
    "\t@staticmethod\n",
    "\tdef __formatAAList(aminoAcids):\n",
    "\t\t\"\"\"Formats 'aminoAcids' into a list of AminoAcid objects.\"\"\"\n",
    "\t\t\n",
    "\t\t#Sequence object's aaList is deep copied\n",
    "\t\tif isinstance(aminoAcids, Sequence):\n",
    "\t\t\treturn deepcopy(aminoAcids._aaList)\n",
    "\t\t\n",
    "\t\t#None becomes an empty Sequence\n",
    "\t\telif aminoAcids is None:\n",
    "\t\t\treturn []\n",
    "\t\t\n",
    "\t\t#A string is converted to a list\n",
    "\t\telif isinstance(aminoAcids, str):\n",
    "\t\t\tif aminoAcids.isupper(): #Multiple Amino Acids in short name mode\n",
    "\t\t\t\treturn [AminoAcid(aa) for aa in aminoAcids]\n",
    "\t\t\telse: #A single Amino Acid with any name mode\n",
    "\t\t\t\treturn [AminoAcid(aminoAcids)]\n",
    "\t\t\n",
    "\t\t#An AminoAcid is copied (from name) and put within a list\n",
    "\t\telif isinstance(aminoAcids, AminoAcid):\n",
    "\t\t\treturn [AminoAcid(aminoAcids)] #Copy constructor\n",
    "\t\t\n",
    "\t\t#A list is copied with all of its items converted to AminoAcid objects\n",
    "\t\telif isinstance(aminoAcids, list):\n",
    "\t\t\treturn [AminoAcid(aa) for aa in aminoAcids]\n",
    "\t\t\n",
    "\t\t#No more supported types\n",
    "\t\telse:\n",
    "\t\t\traise TypeError(\"aminoAcids must be a Sequence, list, AminoAcid object, string or None\")\n",
    "\t\n",
    "\t#Size and comparison\n",
    "\tdef __len__(self):\n",
    "\t\treturn len(self._aaList)\n",
    "\t\t\n",
    "\tdef __gt__(self, other):\n",
    "\t\treturn len(self) > len(other)\n",
    "\t\t\n",
    "\tdef __lt__(self, other):\n",
    "\t\treturn len(self) < len(other)\n",
    "\t\t\n",
    "\tdef __ge__(self, other):\n",
    "\t\treturn len(self) >= len(other)\n",
    "\t\t\n",
    "\tdef __le__(self, other):\n",
    "\t\treturn len(self) <= len(other)\n",
    "\t\n",
    "\tdef __eq__(self, other):\n",
    "\t\treturn len(self) == len(other) and all(a == b for a, b in zip(self, other))\n",
    "\t\n",
    "\tdef __ne__(self, other):\n",
    "\t\treturn not self == other\n",
    "\t\n",
    "\t#Iteration\n",
    "\tdef __iter__(self):\n",
    "\t\treturn iter(self._aaList)\n",
    "\t\n",
    "\t#Representation\n",
    "\tdef __repr__(self):\n",
    "\t\t\"\"\"Representation\"\"\"\n",
    "\t\treturn str(self)\n",
    "\t\t\n",
    "\tdef __str__(self):\n",
    "\t\t\"\"\"String conversion\"\"\"\n",
    "\t\treturn self._separator.join([aa.getName(self._nameMode) for aa in self])\n",
    "\t\t\n",
    "\tdef getDescription(self):\n",
    "\t\t\"\"\"Returns the description of the sequence.\"\"\"\n",
    "\t\treturn self._description\n",
    "\t\n",
    "\tdef setNameMode(self, newMode):\n",
    "\t\t\"\"\"Changes the name display mode to 'newMode'.\"\"\"\n",
    "\t\tif newMode in (\"long\", \"medium\", \"short\"):\n",
    "\t\t\tself._nameMode = newMode\n",
    "\t\telse:\n",
    "\t\t\traise ValueError(\"newMode must be 'long', 'medium' or 'short'\")\n",
    "\t\n",
    "\tdef setSeparator(self, newSep):\n",
    "\t\t\"\"\"Changes the string that separates each displayed AminoAcid.\"\"\"\n",
    "\t\tself._separator = newSep\n",
    "\t\n",
    "\t#Item and slice manipulation\t\n",
    "\tdef __getitem__(self, key):\n",
    "\t\t\"\"\"Return a Sequence object containing copies of the items from the slice\"\"\"\n",
    "\t\tif isinstance(key, slice):\n",
    "\t\t\treturn Sequence([self._aaList[index] for index in range(*key.indices(len(self)))])\n",
    "\t\telse:\n",
    "\t\t\ttry:\n",
    "\t\t\t\treturn self._aaList[key] #Return the item of index 'key'\n",
    "\t\t\texcept:\n",
    "\t\t\t\traise ValueError(\"key does not represent an index or slice\")\n",
    "\t\t\n",
    "\tdef __setitem__(self, key, value):\t\t\n",
    "\t\t\"\"\"Set value for a slice of the sequence\"\"\"\n",
    "\t\tif isinstance(key, slice):\n",
    "\t\t\tfor index in range(*key.indices(len(self))): #range(start, stop, step)\n",
    "\t\t\t\tself._aaList[index] = AminoAcid(value) #create copies\n",
    "\t\telse:\n",
    "\t\t\ttry:\n",
    "\t\t\t\tself._aaList[key] = AminoAcid(value) #Set value for one item\n",
    "\t\t\texcept:\n",
    "\t\t\t\traise ValueError(\"key does not represent an index or slice\")\n",
    "\t\t\n",
    "\tdef __delitem__(self, key):\n",
    "\t\t\"\"\"Delete a slice of the sequence\"\"\"\n",
    "\t\tif isinstance(key, slice):\n",
    "\t\t\tstart, stop, step = key.indices(len(self))\n",
    "\t\t\tdel self._aaList[start:stop:step]\n",
    "\t\telse:\n",
    "\t\t\ttry:\n",
    "\t\t\t\tdel self._aaList[key] #Delete one item\n",
    "\t\t\texcept:\n",
    "\t\t\t\traise ValueError(\"key does not represent an index or slice\")\n",
    "\t\n",
    "\t\n",
    "\t#Sequence modification\n",
    "\tdef insert(self, subSequence, index=0):\n",
    "\t\t\"\"\"\n",
    "\t\tInserts subSequence into the Sequence at index 'index' (default is 0).\n",
    "\t\tsubSequence must be compatible with an AminoAcid list, as specified by __formatAAList.\n",
    "\t\t\"\"\"\n",
    "\t\t#We need a formatted sequence\n",
    "\t\tfor aa in self.__formatAAList(subSequence):\n",
    "\t\t\tself._aaList.insert(index, aa)\n",
    "\t\t\tindex += 1\n",
    "\t\t\t\n",
    "\tdef extend(self, subSequence):\n",
    "\t\t\"\"\"\n",
    "\t\tSame as calling insert at the end of the Sequence.\n",
    "\t\t\"\"\"\n",
    "\t\tself.insert(subSequence, len(self))\n",
    "\t\t\n",
    "\t\t\t\n",
    "\tdef remove(self, subSequence, startIndex=0):\n",
    "\t\t\"\"\"\n",
    "\t\tRemoves the first occurrence of 'subSequence' in self, starting at index. Returns\n",
    "\t\t- start index of the sub-sequence within self (if it exists)\n",
    "\t\t- False if subSequence is not a sub-sequence of self\n",
    "\t\t\"\"\"\n",
    "\t\tlookupResult = self.hasSubSequence(subSequence, startIndex)\n",
    "\t\tif isinstance(lookupResult, int):\n",
    "\t\t\tdel self[startIndex:startIndex+len(subSequence)]\n",
    "\t\t\n",
    "\t\treturn lookupResult\n",
    "\t\n",
    "\t\t\t\n",
    "\tdef delete(self, start=0, stop=None, step=None):\n",
    "\t\t\"\"\"\n",
    "\t\tDeletes Amino Acids between indexes start (included) and stop (excluded).\n",
    "\t\tIf start is not specified, deletion will begin at index 0.\n",
    "\t\tIf stop is not specified, deletion will stop after one item.\n",
    "\t\t\"\"\"\n",
    "\t\tif stop is None:\n",
    "\t\t\tstop = start + 1\n",
    "\t\tif step is None:\n",
    "\t\t\tstep = 1\n",
    "\t\tdel self[start:stop:step]\n",
    "\t\n",
    "\t\n",
    "\t#Lookup\n",
    "\tdef __contains__(self, item):\n",
    "\t\t\"\"\"\n",
    "\t\tReturns True if item is contained in Sequence, False if not.\n",
    "\t\t\"\"\"\n",
    "\t\treturn item in self._aaList\n",
    "\t\n",
    "\tdef hasSubSequence(self, subSequence, startIndex=0):\n",
    "\t\t\"\"\"\n",
    "\t\tChecks if 'subSequence' is a sub-sequence of self, starting at startIndex. Returns\n",
    "\t\t- start index of the sub-sequence within self, if it exists\n",
    "\t\t- False if sequence is not a sub-sequence of self\n",
    "\t\t\"\"\"\n",
    "\t\tsubLen = len(subSequence)\n",
    "\t\tendIndex = len(self)-subLen+1\n",
    "\t\tif endIndex < startIndex or startIndex < 0:\n",
    "\t\t\traise ValueError(\"subSequence does not fit in sequence after startIndex\")\n",
    "\t\t\t\n",
    "\t\tfor i in range(startIndex, endIndex):\n",
    "\t\t\tif all(self[i+j] == subSequence[j] for j in range(subLen)):\n",
    "\t\t\t\treturn i\n",
    "\t\t\n",
    "\t\treturn False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we may not want to type every sequence we use by hand, we can also read them from files. The format chosen here is .fasta, but this can be adapted for any other format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def loadFasta(path):\n",
    "\t\"\"\"\n",
    "\tLoads the fasta file located in 'path' and yields the Sequences it contains.\n",
    "\t\"\"\"\n",
    "\twith open(path, 'r') as fastaFile:\n",
    "\t\tnewSequence = None\n",
    "\t\tfor line in fastaFile:\n",
    "\t\t\tline_s = line.strip()\n",
    "\t\t\tif line_s != \"\" and line_s[0] == \">\":\n",
    "\t\t\t\tif newSequence is not None:\n",
    "\t\t\t\t\tyield newSequence\n",
    "\t\t\t\tnewSequence = Sequence(None, line_s[1:])\n",
    "\t\t\telse:\n",
    "\t\t\t\tnewSequence.extend(line_s)\n",
    "\t\tif len(newSequence)>0:\n",
    "\t\t\tyield newSequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's test a few of the capabilities of this class :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ABX\n",
      "ABXC\n",
      "True\n",
      "lysine-cysteine, lysine-cysteine, ABXC\n",
      "GGVTTFVALYDYESRTETDLSFKKGERLQIVNNTEGDWWLAHSLSTGQTGYIPSNYVAPSDS\n",
      "26\n"
     ]
    }
   ],
   "source": [
    "s = Sequence(\"ABX\")\n",
    "print(s)\n",
    "s.extend(\"cysteine\")\n",
    "print(s)\n",
    "print(AminoAcid(\"alanine\") in s)\n",
    "sCopy = Sequence(s) #This is a deep copy\n",
    "sAlias = s #This is the same object\n",
    "del s[1:3]\n",
    "sAlias[0] = \"K\"\n",
    "s.setSeparator(\"-\")\n",
    "s.setNameMode(\"long\")\n",
    "print(s, sAlias, sCopy, sep=\", \")\n",
    "\n",
    "sequences = [seq for seq in loadFasta(\"SH3-sequences.fasta\")]\n",
    "print(sequences[0])\n",
    "print(sequences[0].hasSubSequence(Sequence(\"RLQIVNNTEGDWWLAHSLS\"))) #Returns the index of the sub-sequence, if found"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Score\n",
    "\n",
    "Now that the easy part is done, we need to remember why we're here (it's sequence alignment, just in case). We can align two sequences any number of ways, but we're only interested in alignments that could represent two descendants of the same original sequence. Therefore, we need to assign a **score** value to each alignment, that will help us see how significative it is.\n",
    "\n",
    "Because of the way mutations happen, all modifications should not be treated equally. Some amino acids are more related between them than others, therefore making them more likely to mutate into each other. Some changes are more or less likely to happen because of the very structure of the protein. Also, when an amino acid is inserted into or deleted from a sequence (thus creating a **gap** in the alignment), it usually is not the only one. This tells us that, not only should the score of the alignment depend on each pair that's aligned (meaning, on the same location), but it should also depend on how many gaps there are and how big they are.\n",
    "\n",
    "The `Score` class allows us to give a numerical value to each possible pair of amino acids. These values can be set manually one by one, just as they can be loaded from files. The chosen format here was `.iij`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "class Score:\n",
    "\t\"\"\"\n",
    "\tRepresents a scoring matrix, used to determine the score between two Amino Acids\n",
    "\t\"\"\"\n",
    "\t\n",
    "\tdef __init__(self, path=\"\", description=\"\", ignore=None):\n",
    "\t\t\"\"\"\n",
    "\t\tCreates a Score object.\n",
    "\t\tIf 'path' is provided, loads the Score values from an iij file.\n",
    "\t\tOtherwise, creates a Score for all possible AminoAcids with values 0.\n",
    "\t\t\"\"\"\n",
    "\t\tself._description = description\n",
    "\t\tself._ignore = Sequence(ignore)\n",
    "\t\tself._matrix = []\n",
    "\t\tself._aaOrder = {}\n",
    "\t\tself._aaSequence = Sequence()\n",
    "\t\t\n",
    "\t\t#If path is provided, load directly from iij file\n",
    "\t\tif path != \"\":\n",
    "\t\t\twith open(path, 'r') as file:\n",
    "\t\t\t\tfoundAAOrder = False #Have we found the line with the amino acid values and order yet?\n",
    "\t\t\t\tfor line in file:\n",
    "\t\t\t\t\tif line[0] != \"#\": #Comments\n",
    "\t\t\t\t\t\t\n",
    "\t\t\t\t\t\tif not foundAAOrder: #Read aa values and order\n",
    "\t\t\t\t\t\t\tfor aa in line.split():\n",
    "\t\t\t\t\t\t\t\tself._aaSequence.extend(aa)\n",
    "\t\t\t\t\t\t\tself._aaOrder = {aa: index for aa, index in zip(self._aaSequence, range(len(self._aaSequence)))}\n",
    "\t\t\t\t\t\t\tfoundAAOrder = True\n",
    "\t\t\t\t\t\telse: #Read matrix values\n",
    "\t\t\t\t\t\t\tself._matrix.append([int(v) for v in line.split()])\n",
    "\t\t\n",
    "\t\t#Otherwise initialize matrix with 0\n",
    "\t\telse:\n",
    "\t\t\tlineSize = 1\n",
    "\t\t\tfor aa in AminoAcid.getAllNames():\n",
    "\t\t\t\tif AminoAcid(aa) not in self._ignore:\n",
    "\t\t\t\t\tself._aaSequence.extend(aa)\n",
    "\t\t\t\t\tself._aaOrder[self._aaSequence[-1]] = lineSize-1\n",
    "\t\t\t\t\tself._matrix.append([0 for i in range(lineSize)])\n",
    "\t\t\t\t\tlineSize += 1\n",
    "\t\t\t\t\n",
    "\t#Representation\n",
    "\tdef __repr__(self):\n",
    "\t\t\"\"\"\n",
    "\t\tRepresentation.\n",
    "\t\t\"\"\"\n",
    "\t\tsepSize = 4\n",
    "\t\tresult = [\"---------- \" + self._description + \" ----------\"]\n",
    "\t\tfor values, aa in zip(self._matrix, self._aaSequence):\n",
    "\t\t\ttempstr = '{a!s:<{w}}'.format(a=aa, w=sepSize)\n",
    "\t\t\tfor value in values:\n",
    "\t\t\t\ttempstr += '{v:<{w}}'.format(v=value, w=sepSize)\n",
    "\t\t\tresult.append(tempstr)\n",
    "\t\ttempstr = \" \"*sepSize\n",
    "\t\tfor aa in self._aaSequence :\n",
    "\t\t\ttempstr += '{a!s:<{w}}'.format(a=aa, w=sepSize)\n",
    "\t\tresult.append(\"\")\n",
    "\t\tresult.append(tempstr)\n",
    "\t\treturn \"\\n\".join(result)\n",
    "\t\n",
    "\t#Scoring\n",
    "\tdef setScore(self, aa1, aa2, score):\n",
    "\t\t\"\"\"\n",
    "\t\tSet the score assigned to AminoAcids 'aa1', 'aa2'.\n",
    "\t\t\"\"\"\n",
    "\t\tid1 = self._aaOrder[aa1]\n",
    "\t\tid2 = self._aaOrder[aa2]\n",
    "\t\tif id1 > id2:\n",
    "\t\t\tself._matrix[id1][id2] = score\n",
    "\t\telse:\n",
    "\t\t\tself._matrix[id2][id1] = score\n",
    "\t\t\n",
    "\tdef getScore(self, aa1, aa2):\n",
    "\t\t\"\"\"\n",
    "\t\tGet the score assigned to AminoAcids 'aa1', 'aa2'.\n",
    "\t\t\"\"\"\n",
    "\t\tid1 = self._aaOrder[aa1]\n",
    "\t\tid2 = self._aaOrder[aa2]\n",
    "\t\tif id1 > id2:\n",
    "\t\t\treturn self._matrix[id1][id2]\n",
    "\t\telse:\n",
    "\t\t\treturn self._matrix[id2][id1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's an example on how to load a `Score` object from a `.iij` file, and access the scores for certains pairs :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------- BLOSUM 62 ----------\n",
      "A   4   \n",
      "R   -1  5   \n",
      "N   -2  0   6   \n",
      "D   -2  -2  1   6   \n",
      "C   0   -3  -3  -3  9   \n",
      "Q   -1  1   0   0   -3  5   \n",
      "E   -1  0   0   2   -4  2   5   \n",
      "G   0   -2  0   -1  -3  -2  -2  6   \n",
      "H   -2  0   1   -1  -3  0   0   -2  8   \n",
      "I   -1  -3  -3  -3  -1  -3  -3  -4  -3  4   \n",
      "L   -1  -2  -3  -4  -1  -2  -3  -4  -3  2   4   \n",
      "K   -1  2   0   -1  -3  1   1   -2  -1  -3  -2  5   \n",
      "M   -1  -1  -2  -3  -1  0   -2  -3  -2  1   2   -1  5   \n",
      "F   -2  -3  -3  -3  -2  -3  -3  -3  -1  0   0   -3  0   6   \n",
      "P   -1  -2  -2  -1  -3  -1  -1  -2  -2  -3  -3  -1  -2  -4  7   \n",
      "S   1   -1  1   0   -1  0   0   0   -1  -2  -2  0   -1  -2  -1  4   \n",
      "T   0   -1  0   -1  -1  -1  -1  -2  -2  -1  -1  -1  -1  -2  -1  1   5   \n",
      "W   -3  -3  -4  -4  -2  -2  -3  -2  -2  -3  -2  -3  -1  1   -4  -3  -2  11  \n",
      "Y   -2  -2  -2  -3  -2  -1  -2  -3  2   -1  -1  -2  -1  3   -3  -2  -2  2   7   \n",
      "V   0   -3  -3  -3  -1  -2  -2  -3  -3  3   1   -2  1   -1  -2  -2  0   -3  -1  4   \n",
      "B   -2  -1  3   4   -3  0   1   -1  0   -3  -4  0   -3  -3  -2  0   -1  -4  -3  -3  4   \n",
      "Z   -1  0   0   1   -3  3   4   -2  0   -3  -3  1   -1  -3  -1  0   -1  -3  -2  -2  1   4   \n",
      "X   0   -1  -1  -1  -2  -1  -1  -1  -1  -1  -1  -1  -1  -1  -2  0   0   -2  -1  -1  -1  -1  -1  \n",
      "\n",
      "    A   R   N   D   C   Q   E   G   H   I   L   K   M   F   P   S   T   W   Y   V   B   Z   X   \n",
      "H N  :  1\n"
     ]
    }
   ],
   "source": [
    "scoring = Score(\"blosum62.iij\", \"BLOSUM 62\")\n",
    "print(scoring)\n",
    "a1, a2 = Sequence(\"HN\") #That's call unpacking, pretty neat huh ?\n",
    "print(a1, a2, \" : \", scoring.getScore(a1, a2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Needleman-Wunsch\n",
    "\n",
    "I know, I know : all these lines and still no alignment in sight. What does that title even mean ? Well, in 1970 Saul B. Needleman and Christian D. Wunsch came up with an effective algorithm for aligning sequences. It provides us with the alignments that get the best score, given certain conditions. It uses a scoring system like the one we've just covered, with the addition of gap penalties : negative values added to the score when the alignment creates (initial penalty) and extends (extended penalty) a gap. Here's an overview of its steps :\n",
    " * Create a matrix with enough rows to fit one sequence ($A$) and enough columns to fit the other ($B$) : each cell $(i, j)$ represents a possible alignment between two amino acids $A_i$ and $B_j$.\n",
    " * Add an initial row and column to the matrix, with values (scores) determined a certain number of ways. Keep in mind that these cells represent the beginning of an alignment where one sequence only has gaps, same with the last row and column for the end of the alignment.\n",
    " * Go through every cell in the matrix and calculate its score based on the previous (left, top, left and top) cells, using the following formula where $(V,W,S)_{i,j}$ are 3 values contained in cell $(i,j)$ of the matrix, $Score(A_i, B_j)$ is the score between amino acids $A_i$ and $B_j$, $I$ is the initial gap penalty and $E$ the extended gap penalty.\n",
    "$$\n",
    "V(i,j) = max\n",
    "\\left\\{\n",
    "    \\begin{array}{ll}\n",
    "      S(i-1, j) + I\\\\\n",
    "      V(i-1, j) - E\n",
    "    \\end{array}\n",
    "\\right.\n",
    "\\quad\n",
    "W(i,j) = max\n",
    "\\left\\{\n",
    "    \\begin{array}{ll}\n",
    "      S(i, j-1) + I\\\\\n",
    "      V(i, j-1) - E\n",
    "    \\end{array}\n",
    "\\right.\n",
    "\\quad\n",
    "S(i,j) = max\n",
    "\\left\\{\n",
    "    \\begin{array}{ll}\n",
    "      S(i-1, j-1) + Score(A_i, B_j)\\\\\n",
    "      V(i, j)\\\\\n",
    "      W(i, j)\n",
    "    \\end{array}\n",
    "\\right.\n",
    "$$\n",
    " * Backtrack from some point of the matrix (the end of the alignment) to some other (the beginning), only passing by permitted cells. The cells allowed after cell $(i,j)$ are the ones where the value $S(i,j)$ comes from : \n",
    "     * left if $S(i,j)=W(i,j)$ : sequence $A$ has a gap\n",
    "     * top if $S(i,j)=V(i,j)$ : sequence $B$ has a gap\n",
    "     * diagonal if $S(i,j)=S(i-1,j-1) + Score(A_i, B_j)$ : sequences $A$ and $B$ are aligned\n",
    "     \n",
    "Different types of alignments can be obtained by tweaking this algorithm.\n",
    " * **Global alignments** aim to align both sequences completely. In order to do that, we initialize the first row and sequences with multiples of $I$ and $E$, thus giving us negative values matching the gap required to get there. Backtracking starts at the end of the matrix and ends at the beginning.\n",
    " * **Local alignments** aim to produce the alignment with the best score, whithout regard for their lenght. We do not initialize the first row and column : completing an alignment with only gaps has no interest score-wise. Backtracking starts at the highest value(s) in the matrix and ends as soon as we reach a value of $0$. Local **suboptimal alignments** can be found by clearing the values of the local optimal alignment in the matrix, reevaluating scores for further rows and columns, and backtracking again.\n",
    " * **Semiglobal alignments** are intended for a global-like alignment between sequences that only overlap partially, or that have great difference in size (one is included in the other). The first row and column are not initialized for the same reason as with local alignments. Backtracking starts and tha highest value(s) but ends when we reach either the first line or first column (therefore finishing one sequence).\n",
    "\n",
    "\n",
    "The following is a class that allows us to represent two aligned sequences, along with information about the way they were aligned and the result. **Identity** is the number of equal amino acids that are aligned, **similarity** is the number of similar (meaning equal or with a non negative score) amino acids that are aligned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "class AlignedSequences:\n",
    "\t\"\"\"\n",
    "\tRepresents two aligned sequences with some metadata about the alignemnt.\n",
    "\t\"\"\"\n",
    "\tdef __init__(self, seqA, seqB, seqInter, score, identity, gaps, similarity, alignmentType):\n",
    "\t\tself.seqA = seqA #Aligned sequences\n",
    "\t\tself.seqB = seqB\n",
    "\t\tself.seqInter = seqInter\n",
    "\t\tself.score = score #Score of alignment\n",
    "\t\tself.identity = identity #Identity\n",
    "\t\tself.gaps = gaps #Gaps count\n",
    "\t\tself.similarity = similarity + identity\n",
    "\t\tself.alignmentType = alignmentType #Options used for alignemnt\n",
    "\t\tself.condensed = False\n",
    "\t\tself.chunkSize = 80 #Used for display\n",
    "\t\t\n",
    "\tdef __repr__(self):\n",
    "\t\t\"\"\"\n",
    "\t\tObject representation.\n",
    "\t\t\"\"\"\n",
    "\t\tres = []\n",
    "\t\tres.append(\"---------- Alignment ----------\")\n",
    "\t\tres.append(\"Type       : \" + self.alignmentType)\n",
    "\t\tres.append(\"Score      : \" + str(self.score))\n",
    "\t\tres.append(\"Identity   : \" + str(self.identity) \\\n",
    "\t\t\t+ \" ({0:.2f}%)\".format(100*(self.identity/len(self.seqA))))\n",
    "\t\tres.append(\"Similarity : \" + str(self.similarity) \\\n",
    "\t\t\t+ \" ({0:.2f}%)\".format(100*(self.similarity/len(self.seqA))))\n",
    "\t\tres.append(\"Gaps       : \" + (str(self.gaps) if self.gaps > 0 else \"None\"))\n",
    "\t\tres.append(\"\")\n",
    "\t\tres.append(\"Upper Sequence : \" + self.seqA.getDescription())\n",
    "\t\tres.append(\"Lower Sequence : \" + self.seqB.getDescription())\n",
    "\t\tres.append(\"\")\n",
    "\t\tif self.condensed:\n",
    "\t\t\treturn \"\\n\".join(res)\n",
    "\t\t\n",
    "\t\tlistA, listB, listI = [], [], []\n",
    "\t\tfor a, i, b in zip(self.seqA, self.seqInter, self.seqB):\n",
    "\t\t\tlistA.append(str(a))\n",
    "\t\t\tlistB.append(str(b))\n",
    "\t\t\tlistI.append(i)\n",
    "\t\t\tif len(listA) == self.chunkSize:\n",
    "\t\t\t\tres.extend([\"\".join(listA), \"\".join(listI), \"\".join(listB), \"\\n\"])\n",
    "\t\t\t\tlistA, listB, listI = [], [], []\n",
    "\t\t\n",
    "\t\tif len(listA) > 0:\n",
    "\t\t\tres.extend([\"\".join(listA), \"\".join(listI), \"\".join(listB), \"\"])\n",
    "\n",
    "\t\treturn \"\\n\".join(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to use this class, we must (finally) implement the actual alignment algorithm. This is done by the following class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "class AlignMatrix:\n",
    "\t\"\"\"\n",
    "\tRepresents an alignment matrix, used to determine the alignment score between two sequences\n",
    "\t\"\"\"\n",
    "\t\n",
    "\tdef __init__(self, scoreMatrix):\n",
    "\t\t\"\"\"\n",
    "\t\tCreates an AlignMatrix object that uses scoreMatrix as a scoring system between amino acids.\n",
    "\t\t\"\"\"\n",
    "\t\tself._scoreMatrix = scoreMatrix #scoring matrix to use\n",
    "\t\t\n",
    "\t\tself._colSeq = None #copies of sequences to align\n",
    "\t\tself._rowSeq = None\n",
    "\t\t\n",
    "\t\tself._alignMatrix = None #Used for alignment scores between sequences\n",
    "\t\tself._rowGapMatrix = None #used for affine gap penalty\n",
    "\t\tself._colGapMatrix = None\n",
    "\t\tself._originMatrix = None #Used for backtracking purposes\n",
    "\t\t\n",
    "\t\tself._alignMode = \"\" #Align mode (global, semiglobal, local)\n",
    "\t\tself._isSuboptimal = False #Is the alignment subobtimal ?\n",
    "\t\t\n",
    "\t\tself._maxAlignScore = 0 #Maximum score\n",
    "\t\tself._currentAlignScore = 0 #Score for this alignment\n",
    "\t\t\n",
    "\t\tself._maxScoreRows = [] #Index of maximum values from _alignMatrix\n",
    "\t\tself._maxScoreCols = []\n",
    "\t\tself._currentAlignPath = []\n",
    "\t\tself._bestAlignPath = [] #sequence of origins for alignment with best score\n",
    "\t\t\n",
    "\t\tself._iniGapPenalty = 0 #Initial gap penalty\n",
    "\t\tself._extGapPenalty = 0 #Extended gap penalty\n",
    "\t\t\n",
    "\t\tself._alignedSeqA = None #Aligned sequences (result of alignment)\n",
    "\t\tself._alignedSeqB = None\n",
    "\t\n",
    "\t\n",
    "\tdef globalAlign(self, seqA, seqB, iniGapPenalty, extGapPenalty=None, semiGlobal=False):\n",
    "\t\t\"\"\"\n",
    "\t\tReturns all best global alignment between 'seqA' and 'seqB' with the provided\n",
    "\t\tinitial and extended gap penalties.\n",
    "\t\tIf 'semiGlobal' is True, allows alignment to discard the start and end of either sequence.\n",
    "\t\t(this allows for better alignments when sequences overlap only partially)\n",
    "\t\t\"\"\"\n",
    "\t\tif semiGlobal:\n",
    "\t\t\tself._alignMode = \"semiglobal\"\n",
    "\t\telse:\n",
    "\t\t\tself._alignMode = \"global\"\n",
    "\t\t\n",
    "\t\t#Initialize all data structures\n",
    "\t\tself.__initialize(seqA, seqB, iniGapPenalty, extGapPenalty)\n",
    "\t\t\n",
    "\t\t#Align in global mode, from the bottom right\n",
    "\t\tif self._alignMode == \"global\":\n",
    "\t\t\tself._currentAlignScore = self._alignMatrix[-1][-1]\n",
    "\t\t\tyield from self.__align(len(self._rowSeq), len(self._colSeq))\n",
    "\t\telse:\n",
    "\t\t\tself._currentAlignScore = self._maxAlignScore\n",
    "\t\t\tfor row, col in zip(self._maxScoreRows, self._maxScoreCols):\n",
    "\t\t\t\tyield from self.__align(row, col)\n",
    "\t\n",
    "\t\n",
    "\tdef localAlign(self, seqA, seqB, iniGapPenalty, extGapPenalty=None, subOptimal=False):\n",
    "\t\t\"\"\"\n",
    "\t\tReturns all best local alignment between 'seqA' and 'seqB' with the provided\n",
    "\t\tinitial and extended gap penalties.\n",
    "\t\tIf 'subOptimal' is True, looks for best suboptimal alignments as well.\n",
    "\t\t\"\"\"\n",
    "\t\tself._alignMode = \"local\"\n",
    "\t\t\n",
    "\t\t#Initialize all data structures\n",
    "\t\tself.__initialize(seqA, seqB, iniGapPenalty, extGapPenalty)\n",
    "\t\t\n",
    "\t\t#Align in local mode, from the maximum value (optimal)\n",
    "\t\tself._currentAlignScore = self._maxAlignScore\n",
    "\t\tfor row, col in zip(self._maxScoreRows, self._maxScoreCols):\n",
    "\t\t\tyield from self.__align(row, col)\n",
    "\t\t\n",
    "\t\t#If suboptimal alignments are requested\n",
    "\t\tif subOptimal:\n",
    "\t\t\tself._isSuboptimal = True\n",
    "\t\t\tself.__clearBestPath() #Reevalute scores with best alignment reset to 0\n",
    "\t\t\tself._currentAlignScore = self._maxAlignScore\n",
    "\t\t\t#Align in local mode, from the new maximum value (sub-optimal)\n",
    "\t\t\tfor row, col in zip(self._maxScoreRows, self._maxScoreCols):\n",
    "\t\t\t\tyield from self.__align(row, col)\n",
    "\t\n",
    "\t\n",
    "\tdef __initialize(self, seqA, seqB, iniGapPenalty, extGapPenalty):\n",
    "\t\t\"\"\"\n",
    "\t\tSets all initial values for required data structures.\n",
    "\t\t\"\"\"\n",
    "\t\t#Sequences\n",
    "\t\tif len(seqA)==0 or len(seqB)==0:\n",
    "\t\t\t\traise ValueError(\"seqA and seqB cannot be empty\")\n",
    "\t\tself._colSeq = Sequence(seqA)\n",
    "\t\tself._rowSeq = Sequence(seqB)\n",
    "\t\t\n",
    "\t\t#Alignments\n",
    "\t\tself._subOptimal = False\n",
    "\t\tself._alignSeqA = Sequence()\n",
    "\t\tself._alignSeqB = Sequence()\n",
    "\t\tself._alignInter = \"\"\n",
    "\t\t\n",
    "\t\tself._maxAlignScore = 0 #Maximum score found while aligning\n",
    "\t\tself._maxScoreRows = [] #Rows of maximum score\n",
    "\t\tself._maxScoreCols = [] #Columns of maximum score \n",
    "\t\t\n",
    "\t\tself._currentAlignPath = [] #indexes of the current alignment\n",
    "\t\tself._bestAlignPath = [] #indexes of the alignment with the best score\n",
    "\t\t\n",
    "\t\t#Gap penalties\n",
    "\t\tself._iniGapPenalty = iniGapPenalty\n",
    "\t\tself._extGapPenalty = iniGapPenalty if extGapPenalty is None else extGapPenalty\n",
    "\t\t\n",
    "\t\t#Matrices\n",
    "\t\tself._alignMatrix = [[0 for i in range(len(self._colSeq)+1)] \\\n",
    "\t\t\t\t\t\t\t\tfor j in range(len(self._rowSeq)+1)]\n",
    "\t\tself._rowGapMatrix = deepcopy(self._alignMatrix)\n",
    "\t\tself._colGapMatrix = deepcopy(self._alignMatrix)\n",
    "\t\t\n",
    "\t\tself._originMatrix = [[\"\" for i in range(len(self._colSeq)+1)] \\\n",
    "\t\t\t\t\t\t\t\tfor j in range(len(self._rowSeq)+1)]\n",
    "\t\t\n",
    "\t\t#Global alignment : first line and colunm have initial scores and origins\n",
    "\t\tif self._alignMode== \"global\":\n",
    "\t\t\tfor i in range(1, max(len(self._colSeq), len(self._rowSeq))+1):\t\n",
    "\t\t\t\tval = self._iniGapPenalty + self._extGapPenalty*(i-1)\n",
    "\t\t\t\tif i <= len(self._colSeq): #First row\n",
    "\t\t\t\t\tself._alignMatrix[0][i] = val\n",
    "\t\t\t\t\tself._rowGapMatrix[0][i] = val \n",
    "\t\t\t\t\tself._colGapMatrix[0][i] = val\n",
    "\t\t\t\t\tself._originMatrix[0][i] = \"L\" #Origin is left\n",
    "\t\t\t\t\n",
    "\t\t\t\tif i <= len(self._rowSeq): #First column\n",
    "\t\t\t\t\tself._alignMatrix[i][0] = val\n",
    "\t\t\t\t\tself._rowGapMatrix[i][0] = val \n",
    "\t\t\t\t\tself._colGapMatrix[i][0] = val\n",
    "\t\t\t\t\tself._originMatrix[i][0] = \"T\" #Origin is top\t\t\t\t\t\n",
    "\t\t\t\n",
    "\t\t\n",
    "\t\t#Fill all matrices\n",
    "\t\tfor row in range(1, len(self._rowSeq)+1):\n",
    "\t\t\tfor col in range(1, len(self._colSeq)+1):\n",
    "\t\t\t\tself.__fill(row, col)\n",
    "\t\n",
    "\t\n",
    "\tdef __fill(self, row, col):\n",
    "\t\t\"\"\"\n",
    "\t\tCalculates the score for value at row 'row' and column 'col'.\n",
    "\t\t\"\"\"\n",
    "\t\t#Row gap matrix\n",
    "\t\tinitialGapScore = self._alignMatrix[row-1][col] + self._iniGapPenalty\n",
    "\t\textendedGapScore = self._rowGapMatrix[row-1][col] + self._extGapPenalty\n",
    "\t\tself._rowGapMatrix[row][col] = max(initialGapScore, extendedGapScore)\n",
    "\t\t\n",
    "\t\t#Column gap matrix\n",
    "\t\tinitialGapScore = self._alignMatrix[row][col-1] + self._iniGapPenalty\n",
    "\t\textendedGapScore = self._colGapMatrix[row][col-1] + self._extGapPenalty\n",
    "\t\tself._colGapMatrix[row][col] = max(initialGapScore, extendedGapScore)\n",
    "\t\t\n",
    "\t\t#Align matrix\n",
    "\t\tdeleteScore = self._rowGapMatrix[row][col]\n",
    "\t\tinsertScore = self._colGapMatrix[row][col]\n",
    "\t\tmatchScore = self._alignMatrix[row-1][col-1] + \\\n",
    "\t\t\tself._scoreMatrix.getScore(self._rowSeq[row-1], self._colSeq[col-1])\n",
    "\t\t\t\n",
    "\t\tmaxScore = 0\n",
    "\t\tif self._alignMode == \"local\": #Local alignment\n",
    "\t\t\tself._alignMatrix[row][col] = maxScore = max(insertScore, deleteScore, matchScore, 0)\n",
    "\t\telse:\t#Global alignment\n",
    "\t\t\tself._alignMatrix[row][col] = maxScore = max(insertScore, deleteScore, matchScore)\n",
    "\t\t\t\n",
    "\t\t#Max alignment scores\n",
    "\t\tif maxScore > self._maxAlignScore: #New max alignment\n",
    "\t\t\tself._maxAlignScore = maxScore\n",
    "\t\t\tself._maxScoreRows = [row]\n",
    "\t\t\tself._maxScoreCols = [col]\n",
    "\t\telif maxScore == self._maxAlignScore: #Same max alignment\n",
    "\t\t\tself._maxScoreRows.append(row)\n",
    "\t\t\tself._maxScoreCols.append(col)\n",
    "\t\t\t\n",
    "\t\t#Origin matrix\n",
    "\t\tif insertScore == maxScore:\n",
    "\t\t\tself._originMatrix[row][col] += \"L\"\n",
    "\t\tif matchScore == maxScore:\n",
    "\t\t\tself._originMatrix[row][col] += \"D\"\n",
    "\t\tif deleteScore == maxScore:\n",
    "\t\t\tself._originMatrix[row][col] += \"T\"\n",
    "\t\n",
    "\t\n",
    "\tdef __clearBestPath(self):\n",
    "\t\t\"\"\"\n",
    "\t\tResets all scores from the first best alignment and then reevaluates the affected scores.\n",
    "\t\tA new maximum score will be found from these, allowing for new alignment lookups.\n",
    "\t\t\"\"\"\n",
    "\t\tself._maxAlignScore = 0 #New maximum will be found from modified values\n",
    "\t\tself._maxScoreRows = []\n",
    "\t\tself._maxScoreCols = []\n",
    "\t\t\n",
    "\t\tfor row, col in self._bestAlignPath[::-1]: #Reverse best align path\n",
    "\t\t\tself._alignMatrix[row][col] = 0 #Reset scores\n",
    "\t\t\tself._colGapMatrix[row][col] = 0\n",
    "\t\t\tself._rowGapMatrix[row][col] = 0\n",
    "\t\t\t\n",
    "\t\t\t#Reevaluate scores for further rows and columns\n",
    "\t\t\tfor furtherRow in range(row+1, len(self._rowSeq)+1):\n",
    "\t\t\t\tif (furtherRow, col) not in self._bestAlignPath: #Only if not part of best path\n",
    "\t\t\t\t\tself._originMatrix[furtherRow][col] = \"\"\n",
    "\t\t\t\t\tself.__fill(furtherRow, col)\n",
    "\t\t\t\n",
    "\t\t\tfor furtherCol in range(col+1, len(self._colSeq)+1):\n",
    "\t\t\t\tif (row, furtherCol) not in self._bestAlignPath:\n",
    "\t\t\t\t\tself._originMatrix[row][furtherCol] = \"\"\n",
    "\t\t\t\t\tself.__fill(row, furtherCol)\n",
    "\t\t\t\t\t\n",
    "\t\t\t\n",
    "\t\t\n",
    "\t\n",
    "\tdef __align(self, i, j, identity=0, gaps=0, similarity=0):\n",
    "\t\t\"\"\"\n",
    "\t\tYields all best alignments starting from row i and column j.\n",
    "\t\t\"\"\"\n",
    "\t\t#Local alignments are complete when we reach a null score\n",
    "\t\t#Global alignments are complete when we reach the beginning of the matrix\n",
    "\t\t#Semiglobal alignments are complete when we reach an edge of the matrix\n",
    "\t\tif (self._alignMode == \"local\" and self._alignMatrix[i][j]==0) \\\n",
    "\t\tor (self._alignMode == \"global\" and i==0 and j==0) \\\n",
    "\t\tor (self._alignMode == \"semiglobal\" and (i==0 or j==0)):\n",
    "\t\t\t\n",
    "\t\t\t#Create AlignedSequences object as result\n",
    "\t\t\tresult = AlignedSequences(Sequence(self._alignSeqA, self._colSeq.getDescription()), \\\n",
    "\t\t\tSequence(self._alignSeqB, self._rowSeq.getDescription()),self._alignInter[::-1], \\\n",
    "\t\t\tself._currentAlignScore, identity, gaps, similarity, \\\n",
    "\t\t\tself._alignMode + \"-suboptimal\"*self._isSuboptimal)\n",
    "\t\t\t\n",
    "\t\t\t#Remember first best alignment for subobtimal lookup\n",
    "\t\t\tif self._bestAlignPath == []:\n",
    "\t\t\t\tself._bestAlignPath = deepcopy(self._currentAlignPath)\n",
    "\t\t\t\n",
    "\t\t\tyield result\n",
    "\t\t\t\n",
    "\t\telse:\n",
    "\t\t\tfor origin in self._originMatrix[i][j]:\n",
    "\t\t\t\tself._currentAlignPath.append((i,j))\n",
    "\t\t\t\t\n",
    "\t\t\t\tif origin == \"T\": #top\n",
    "\t\t\t\t\tself._alignSeqA.insert(\"gap\", 0)\n",
    "\t\t\t\t\tself._alignSeqB.insert(self._rowSeq[i-1], 0)\n",
    "\t\t\t\t\tself._alignInter += \" \"\n",
    "\t\t\t\t\tyield from self.__align(i-1, j, identity, gaps+1, similarity)\n",
    "\t\t\t\t\t\n",
    "\t\t\t\telif origin == \"D\": #diagonal\n",
    "\t\t\t\t\tif self._colSeq[j-1] == self._rowSeq[i-1]:\n",
    "\t\t\t\t\t\tidentity += 1\n",
    "\t\t\t\t\t\tself._alignInter += \":\"\n",
    "\t\t\t\t\telif self._scoreMatrix.getScore(self._colSeq[j-1], self._rowSeq[i-1])>=0:\n",
    "\t\t\t\t\t\tsimilarity += 1\n",
    "\t\t\t\t\t\tself._alignInter += \".\"\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\tself._alignInter += \" \"\n",
    "\t\t\t\t\tself._alignSeqA.insert(self._colSeq[j-1], 0)\n",
    "\t\t\t\t\tself._alignSeqB.insert(self._rowSeq[i-1], 0)\n",
    "\t\t\t\t\tyield from self.__align(i-1, j-1, identity, gaps, similarity)\n",
    "\t\t\t\t\n",
    "\t\t\t\telif origin == \"L\": #left\n",
    "\t\t\t\t\tself._alignSeqA.insert(self._colSeq[j-1], 0)\n",
    "\t\t\t\t\tself._alignSeqB.insert(\"gap\", 0)\n",
    "\t\t\t\t\tself._alignInter += \" \"\n",
    "\t\t\t\t\tyield from self.__align(i, j-1, identity, gaps+1, similarity)\n",
    "\t\t\t\t\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\traise ValueError(\"Origin must be T (top), D (diagonal) or L (left)\")\n",
    "\t\t\t\t\n",
    "\t\t\t\tself._currentAlignPath.pop()\n",
    "\t\t\t\tself._alignSeqA.delete(0)\n",
    "\t\t\t\tself._alignSeqB.delete(0)\n",
    "\t\t\t\tself._alignInter = self._alignInter[:-1]\n",
    "\n",
    "\t\t\n",
    "\t#Representation\n",
    "\tdef __repr__(self):\n",
    "\t\t\"\"\"\n",
    "\t\tRepresentation of the alignemnt matrix.\n",
    "\t\t\"\"\"\n",
    "\t\tif self._colSeq is None:\n",
    "\t\t\treturn \"No alignment done yet\"\n",
    "\t\t\n",
    "\t\tresult = []\n",
    "\t\tsepSize = 5\n",
    "\t\trep = {\"\":\"\", \"L\": \"_   \", \"D\": \" \\\\  \", \"T\": \"   |\", \"LD\": \"_\\\\  \",\"DT\": \" \\\\ |\", \"LT\": \"_  |\", \"LDT\": \"_\\\\ |\"}\n",
    "\t\t#Amino Acid column sequence\n",
    "\t\tresult.append(\" \"*2*sepSize)\n",
    "\t\tfor aa in self._colSeq :\n",
    "\t\t\tresult[-1] += '{a!s:<{w}}'.format(a=aa, w=sepSize)\n",
    "\t\t\n",
    "\t\t#First line of values (no amino acid)\n",
    "\t\tresult.append(\" \")\n",
    "\t\tfor origin in self._originMatrix[0]:\n",
    "\t\t\tresult[-1] += '{o:>{w}}'.format(o=rep[origin], w=sepSize)\n",
    "\t\t\n",
    "\t\tresult.append(\" \"*sepSize)\n",
    "\t\tfor value in self._alignMatrix[0]:\n",
    "\t\t\tresult[-1] += '{v:<{w}}'.format(v=value, w=sepSize)\n",
    "\t\t\n",
    "\t\t#Amino Acid line sequence and values\n",
    "\t\tfor values, origins, aa in zip(self._alignMatrix[1:], self._originMatrix[1:], self._rowSeq):\n",
    "\t\t\tresult.append(\" \")\n",
    "\t\t\tfor origin in origins:\n",
    "\t\t\t\tresult[-1] += '{o:>{w}}'.format(o=rep[origin], w=sepSize)\n",
    "\t\t\tresult.append(\"\")\n",
    "\t\t\tresult[-1] += '{a!s:<{w}}'.format(a=aa, w=sepSize)\n",
    "\t\t\tfor value in values:\n",
    "\t\t\t\tresult[-1] += '{v:<{w}}'.format(v=value, w=sepSize)\n",
    "\t\t\n",
    "\t\treturn \"\\n\".join(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is the result of a global alignment between the first two sequences from \"SH3-sequences.fasta\", calculated by LALIGN with the BLOSUM62 scoring matrix, initial and extended gap penalties of $-12$ and $-2$ :\n",
    "```\n",
    "n-w opt:  69  Z-score: 152.4  bits: 31.6 E(1): 6.7e-25\n",
    "global/global (N-W) score: 69; 29.0% identity (62.9% similar) in 62 aa overlap (1-62:1-58)\n",
    "\n",
    "GGVTTFVALYDYESRTETDLSFKKGERLQIVNNTEGDWWLAHSLSTGQTGYIPSNYVAPSDS\n",
    "      .: ::... .. .::::.:. :...:.   . :   .:. :. :.::.::.  .\n",
    "---MEAIAKYDFKATADDELSFKRGDILKVLNEECDQNWYKAELN-GKDGFIPKNYIEMKPH\n",
    "```\n",
    "\n",
    "Next is the result from the AlignMatrix class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------- Alignment ----------\n",
      "Type       : global\n",
      "Score      : 73\n",
      "Identity   : 18 (29.03%)\n",
      "Similarity : 39 (62.90%)\n",
      "Gaps       : 4\n",
      "\n",
      "Upper Sequence : sp|P12931|84-145\n",
      "Lower Sequence : sp|P62993|1-58\n",
      "\n",
      "GGVTTFVALYDYESRTETDLSFKKGERLQIVNNTEGDWWLAHSLSTGQTGYIPSNYVAPSDS\n",
      "      .: ::... .. .::::.:. :...:.   . :   .:. :. :.::.::.  .  \n",
      "---MEAIAKYDFKATADDELSFKRGDILKVLNEECDQNWYKAELN-GKDGFIPKNYIEMKPH\n",
      "\n"
     ]
    }
   ],
   "source": [
    "s = Score(\"blosum62.iij\", \"BLOSUM 62\")\n",
    "a = AlignMatrix(s)\n",
    "sequences = [seq for seq in loadFasta(\"SH3-sequences.fasta\")]\n",
    "\n",
    "for align in a.globalAlign(sequences[0], sequences[1], -12, -2, False):\n",
    "\tprint(align)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Besides the second one being much more readable, their output is pretty similar. There may be slight differences in the BLOSUM matrix used, responsible for the discrepancy between the scores.\n",
    "\n",
    "Here is the result of a local alignment between the first two sequences from \"maguk-sequences.fasta\", calculated by LALIGN with the BLOSUM62 scoring matrix, initial and extended gap penalties of $-12$ and $-2$ :\n",
    "```\n",
    "Waterman-Eggert score: 2677;  1048.2 bits; E(1) <  0\n",
    "69.5% identity (86.3% similar) in 767 aa overlap (153-904:61-817)\n",
    "\n",
    "SHSHISPIKPTEA-VLPSPPTVPVIPVLPVPAENTVILP-TIPQANPPPVLVNTDSLETP\n",
    ":..  .:   :.: ..:.   :  .:  :::...:  :  . :.  :  . .: :  :  \n",
    "SQAGATPTPRTKAKLIPTGRDVGPVPPKPVPGKSTPKLNGSGPSWWPECTCTNRDWYEQ-\n",
    "\n",
    "TYVNGTDADYEYEEITLERGNSGLGFSIAGGTDNPHIGDDSSIFITKIITGGAAAQDGRL\n",
    "  :::.:. ..::::.::::::::::::::: ::::. :: .::::::: :::::.::::\n",
    "--VNGSDGMFKYEEIVLERGNSGLGFSIAGGIDNPHVPDDPGIFITKIIPGGAAAMDGRL\n",
    "\n",
    "RVNDCILRVNEVDVRDVTHSKAVEALKEAGSIVRLYVKRRKPVSEKIMEIKLIKGPKGLG\n",
    " ::::.:::::::: .:.::.::::::::: .::: :.::.:  : :::..:.:::::::\n",
    "GVNDCVLRVNEVDVSEVVHSRAVEALKEAGPVVRLVVRRRQPPPETIMEVNLLKGPKGLG\n",
    "\n",
    "FSIAGGVGNQHIPGDNSIYVTKIIEGGAAHKDGKLQIGDKLLAVNNVCLEEVTHEEAVTA\n",
    "::::::.::::::::::::.:::::::::.:::.:::::.::::::. :..: :::::..\n",
    "FSIAGGIGNQHIPGDNSIYITKIIEGGAAQKDGRLQIGDRLLAVNNTNLQDVRHEEAVAS\n",
    "      \n",
    "LKNTSDFVYLKVAKPTSMYMNDGYAPPDITNSSSQPVDNHVSPSSFLGQTPA--------\n",
    "::::::.:::::::: :...:: ::::: ... .  .:::.: .: ::   :        \n",
    "LKNTSDMVYLKVAKPGSLHLNDMYAPPDYASTFTALADNHISHNSSLGYLGAVESKVSYP\n",
    "     \n",
    "-----SPARYSPVSKAVLGDDEITREPRKVVLHRGSTGLGFNIVGGEDGEGIFISFILAG\n",
    "      :.::::. . .:.....::::::..::.:::::::::::::::::::.::::::\n",
    "APPQVPPTRYSPIPRHMLAEEDFTREPRKIILHKGSTGLGFNIVGGEDGEGIFVSFILAG\n",
    "  \n",
    "GPADLSGELRKGDRIISVNSVDLRAASHEQAAAALKNAGQAVTIVAQYRPEEYSRFEAKI\n",
    "::::::::::.::::.:::.:.:: :.:::::::::.:::.::::::::::::::::.::\n",
    "GPADLSGELRRGDRILSVNGVNLRNATHEQAAAALKRAGQSVTIVAQYRPEEYSRFESKI\n",
    "    \n",
    "HDLREQMMNSSISSGSGSLRTSQKRSLYVRALFDYDKTKDSGLPSQGLNFKFGDILHVIN\n",
    ":::::::::::.::::::::::.:::::::::::::.:.:: ::::::.:..::::::::\n",
    "HDLREQMMNSSMSSGSGSLRTSEKRSLYVRALFDYDRTRDSCLPSQGLSFSYGDILHVIN\n",
    "   \n",
    "ASDDEWWQARQVTPDGESDEVGVIPSKRRVEKKERARLKTVKFNSKTRDKGEIPDDMGSK\n",
    ":::::::::: ::: :::...::::::.:::::::::::::::...:   : : ..    \n",
    "ASDDEWWQARLVTPHGESEQIGVIPSKKRVEKKERARLKTVKFHART---GMIESNRDFP\n",
    "   \n",
    "GLKHVTSNASDSESSYRGQEEYVLSYEPVNQQEVNYTRPVIILGPMKDRINDDLISEFPD\n",
    "::    :.   . .. .:::. .::::::..::..:.::::::::::::.::::::::: \n",
    "GL----SDDYYGAKNLKGQEDAILSYEPVTRQEIHYARPVIILGPMKDRVNDDLISEFPH\n",
    "  \n",
    "KFGSCVPHTTRPKRDYEVDGRDYHFVTSREQMEKDIQEHKFIEAGQYNNHLYGTSVQSVR\n",
    "::::::::::::.:: ::::.:::::.::::::::::..:::::::.:..:::::.::::\n",
    "KFGSCVPHTTRPRRDNEVDGQDYHFVVSREQMEKDIQDNKFIEAGQFNDNLYGTSIQSVR\n",
    "    \n",
    "EVAEKGKHCILDVSGNAIKRLQIAQLYPISIFIKPKSMENIMEMNKRLTEEQARKTFERA\n",
    " :::.::::::::::::::::: ::::::.:::::::.: .::::.: : :::.: ...:\n",
    "AVAERGKHCILDVSGNAIKRLQQAQLYPIAIFIKPKSIEALMEMNRRQTYEQANKIYDKA\n",
    " \n",
    "MKLEQEFTEHFTAIVQGDTLEDIYNQVKQIIEEQSGSYIWVPAKEKL\n",
    "::::::: :.::::::::.::.:::..:::::.::: :::::. :::\n",
    "MKLEQEFGEYFTAIVQGDSLEEIYNKIKQIIEDQSGHYIWVPSPEKL  \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 403,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------- Alignment ----------\n",
      "Type       : local\n",
      "Score      : 2689\n",
      "Identity   : 534 (69.62%)\n",
      "Similarity : 663 (86.44%)\n",
      "Gaps       : 25\n",
      "\n",
      "Upper Sequence : sp|Q12959|1-904\n",
      "Lower Sequence : sp|Q92796|DLG3_HUMAN Disks large homolog 3 OS=Homo sapiens GN=DLG3 PE=1 SV=2\n",
      "\n",
      "SHSHISPIKPTEA-VLPSPPTVPVIPVLPVPAENTVIL-PTIPQANPPPVLVNTDSLETP\n",
      ":..  .:   :.: ..:.   :  .:  :::...:  :  . :.  :  . .: :  :  \n",
      "SQAGATPTPRTKAKLIPTGRDVGPVPPKPVPGKSTPKLNGSGPSWWPECTCTNRDWYEQ-\n",
      "\n",
      "\n",
      "TYVNGTDADYEYEEITLERGNSGLGFSIAGGTDNPHIGDDSSIFITKIITGGAAAQDGRL\n",
      "  :::.:. ..::::.::::::::::::::: ::::. :: .::::::: :::::.::::\n",
      "--VNGSDGMFKYEEIVLERGNSGLGFSIAGGIDNPHVPDDPGIFITKIIPGGAAAMDGRL\n",
      "\n",
      "\n",
      "RVNDCILRVNEVDVRDVTHSKAVEALKEAGSIVRLYVKRRKPVSEKIMEIKLIKGPKGLG\n",
      " ::::.:::::::: .:.::.::::::::: .::: :.::.:  : :::..:.:::::::\n",
      "GVNDCVLRVNEVDVSEVVHSRAVEALKEAGPVVRLVVRRRQPPPETIMEVNLLKGPKGLG\n",
      "\n",
      "\n",
      "FSIAGGVGNQHIPGDNSIYVTKIIEGGAAHKDGKLQIGDKLLAVNNVCLEEVTHEEAVTA\n",
      "::::::.::::::::::::.:::::::::.:::.:::::.::::::. :..: :::::..\n",
      "FSIAGGIGNQHIPGDNSIYITKIIEGGAAQKDGRLQIGDRLLAVNNTNLQDVRHEEAVAS\n",
      "\n",
      "\n",
      "LKNTSDFVYLKVAKPTSMYMNDGYAPPDITNSSSQPVDNHVS-PSS--FLG---------\n",
      "::::::.:::::::: :...:: ::::: ... .  .:::.:  ::  .::         \n",
      "LKNTSDMVYLKVAKPGSLHLNDMYAPPDYASTFTALADNHISHNSSLGYLGAVESKVSYP\n",
      "\n",
      "\n",
      "QTPA-SPARYSPVSKAVLGDDEITREPRKVVLHRGSTGLGFNIVGGEDGEGIFISFILAG\n",
      "  :   :.::::. . .:.....::::::..::.:::::::::::::::::::.::::::\n",
      "APPQVPPTRYSPIPRHMLAEEDFTREPRKIILHKGSTGLGFNIVGGEDGEGIFVSFILAG\n",
      "\n",
      "\n",
      "GPADLSGELRKGDRIISVNSVDLRAASHEQAAAALKNAGQAVTIVAQYRPEEYSRFEAKI\n",
      "::::::::::.::::.:::.:.:: :.:::::::::.:::.::::::::::::::::.::\n",
      "GPADLSGELRRGDRILSVNGVNLRNATHEQAAAALKRAGQSVTIVAQYRPEEYSRFESKI\n",
      "\n",
      "\n",
      "HDLREQMMNSSISSGSGSLRTSQKRSLYVRALFDYDKTKDSGLPSQGLNFKFGDILHVIN\n",
      ":::::::::::.::::::::::.:::::::::::::.:.:: ::::::.:..::::::::\n",
      "HDLREQMMNSSMSSGSGSLRTSEKRSLYVRALFDYDRTRDSCLPSQGLSFSYGDILHVIN\n",
      "\n",
      "\n",
      "ASDDEWWQARQVTPDGESDEVGVIPSKRRVEKKERARLKTVKFNSKTRDKGEIPDDMGSK\n",
      ":::::::::: ::: :::...::::::.:::::::::::::::...:   : : ..    \n",
      "ASDDEWWQARLVTPHGESEQIGVIPSKKRVEKKERARLKTVKFHART---GMIESNRDFP\n",
      "\n",
      "\n",
      "GLKHVTSNASDSESSYRGQEEYVLSYEPVNQQEVNYTRPVIILGPMKDRINDDLISEFPD\n",
      "::    :.   . .. .:::. .::::::..::..:.::::::::::::.::::::::: \n",
      "GL----SDDYYGAKNLKGQEDAILSYEPVTRQEIHYARPVIILGPMKDRVNDDLISEFPH\n",
      "\n",
      "\n",
      "KFGSCVPHTTRPKRDYEVDGRDYHFVTSREQMEKDIQEHKFIEAGQYNNHLYGTSVQSVR\n",
      "::::::::::::.:: ::::.:::::.::::::::::..:::::::.:..:::::.::::\n",
      "KFGSCVPHTTRPRRDNEVDGQDYHFVVSREQMEKDIQDNKFIEAGQFNDNLYGTSIQSVR\n",
      "\n",
      "\n",
      "EVAEKGKHCILDVSGNAIKRLQIAQLYPISIFIKPKSMENIMEMNKRLTEEQARKTFERA\n",
      " :::.::::::::::::::::: ::::::.:::::::.: .::::.: : :::.: ...:\n",
      "AVAERGKHCILDVSGNAIKRLQQAQLYPIAIFIKPKSIEALMEMNRRQTYEQANKIYDKA\n",
      "\n",
      "\n",
      "MKLEQEFTEHFTAIVQGDTLEDIYNQVKQIIEEQSGSYIWVPAKEKL\n",
      "::::::: :.::::::::.::.:::..:::::.::: :::::. :::\n",
      "MKLEQEFGEYFTAIVQGDSLEEIYNKIKQIIEDQSGHYIWVPSPEKL\n",
      "\n"
     ]
    }
   ],
   "source": [
    "s = Score(\"blosum62.iij\", \"BLOSUM 62\")\n",
    "a = AlignMatrix(s)\n",
    "sequences = [seq for seq in loadFasta(\"maguk-sequences.fasta\")]\n",
    "\n",
    "for align in a.localAlign(sequences[0], sequences[1], -12, -2, False):\n",
    "\talign.chunkSize = 60\n",
    "\tprint(align)\n",
    "\tbreak"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once more, alignments are quite similar in terms of gap locations, scores and identity/similarity. The results suggest that these sequences might be related, given their high identity and similarity.\n",
    "\n",
    "For each other pair of sequences (condensed results), we don't get such high changes of homology, as suggested by these condensed results :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 404,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------- Alignment ----------\n",
      "Type       : local\n",
      "Score      : 2689\n",
      "Identity   : 534 (69.62%)\n",
      "Similarity : 663 (86.44%)\n",
      "Gaps       : 25\n",
      "\n",
      "Upper Sequence : sp|Q12959|1-904\n",
      "Lower Sequence : sp|Q92796|DLG3_HUMAN Disks large homolog 3 OS=Homo sapiens GN=DLG3 PE=1 SV=2\n",
      "\n",
      "---------- Alignment ----------\n",
      "Type       : local\n",
      "Score      : 369\n",
      "Identity   : 113 (26.84%)\n",
      "Similarity : 263 (62.47%)\n",
      "Gaps       : 26\n",
      "\n",
      "Upper Sequence : sp|Q12959|1-904\n",
      "Lower Sequence : sp|O14936|1-926\n",
      "\n",
      "---------- Alignment ----------\n",
      "Type       : local\n",
      "Score      : 168\n",
      "Identity   : 33 (36.26%)\n",
      "Similarity : 65 (71.43%)\n",
      "Gaps       : 5\n",
      "\n",
      "Upper Sequence : sp|Q12959|1-904\n",
      "Lower Sequence : sp|Q86UL8|1-1455\n",
      "\n",
      "---------- Alignment ----------\n",
      "Type       : local\n",
      "Score      : 428\n",
      "Identity   : 117 (28.19%)\n",
      "Similarity : 259 (62.41%)\n",
      "Gaps       : 21\n",
      "\n",
      "Upper Sequence : sp|Q92796|DLG3_HUMAN Disks large homolog 3 OS=Homo sapiens GN=DLG3 PE=1 SV=2\n",
      "Lower Sequence : sp|O14936|1-926\n",
      "\n",
      "---------- Alignment ----------\n",
      "Type       : local\n",
      "Score      : 168\n",
      "Identity   : 50 (29.24%)\n",
      "Similarity : 102 (59.65%)\n",
      "Gaps       : 8\n",
      "\n",
      "Upper Sequence : sp|Q92796|DLG3_HUMAN Disks large homolog 3 OS=Homo sapiens GN=DLG3 PE=1 SV=2\n",
      "Lower Sequence : sp|Q86UL8|1-1455\n",
      "\n",
      "---------- Alignment ----------\n",
      "Type       : local\n",
      "Score      : 106\n",
      "Identity   : 38 (29.69%)\n",
      "Similarity : 75 (58.59%)\n",
      "Gaps       : 10\n",
      "\n",
      "Upper Sequence : sp|O14936|1-926\n",
      "Lower Sequence : sp|Q86UL8|1-1455\n",
      "\n"
     ]
    }
   ],
   "source": [
    "s = Score(\"blosum62.iij\", \"BLOSUM 62\")\n",
    "a = AlignMatrix(s)\n",
    "sequences = [seq for seq in loadFasta(\"maguk-sequences.fasta\")]\n",
    "\n",
    "for i in range(len(sequences)-1):\n",
    "\tfor j in range(i+1, len(sequences)):\n",
    "\t\tfor align in a.localAlign(sequences[i], sequences[j], -12, -2, False):\n",
    "\t\t\talign.condensed = True\n",
    "\t\t\tprint(align)\n",
    "\t\t\tbreak"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
