{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from aminoacid import AminoAcid\n",
    "from sequence import Sequence, loadFasta\n",
    "from align import Align, Aligned\n",
    "from math import sqrt, log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Project 3 - PSSM Profiles and alignment\n",
    "\n",
    "## Domain\n",
    "\n",
    "A protein **domain** is part of a protein that has its own structure and function, independent of the rest of the protein. Despite their use being similar across the proteins they can be found in, the same domains do not necessarily have the same amino acid sequence. However, they do have enough common amino acids to be identified and located by the means of *sequence alignment*. Our goal in this project is to locate a given domain within a given sequence, should it exist one or more times within that sequence. One such domain is the WW domain, known to be used by multiple species and sometimes included multiple times in a single protein. In this project we will use sequences from the WW domain that belong to human proteins to test our implementations : these have been saved to the file `to-be-aligned.fasta`. These domain sequences and many others can be found on the [SMART](http://smart.embl.de) database.\n",
    "\n",
    "## MSA\n",
    "\n",
    "*Multiple Sequence Alignment* allows us to align multiple sequences together, and therefore observe how well some amino acids are preserved at some positions. This is of great interest in the study of domains, since it reveals which positions are of greater importance when trying to align new sequences to known domains. There are of course other applications, as well as numerous methods that achieve MSA, however we won't go into them since we won't be implementing any of these in this project. Two online tools were used to align all the sequences from `to-be-aligned.fasta` together : [MUSCLE](http://www.ebi.ac.uk/Tools/msa/muscle/) and [CLUSTAL Omega](http://www.ebi.ac.uk/Tools/msa/clustalo/). The resulting sequences can be respectively found in files `msaresults-MUSCLE.fasta` and `msaresults-CLUSTALO.fasta`.\n",
    "\n",
    "## PSSM\n",
    "\n",
    "A *profile* is a way to represent multiple aligned sequences and their similarities or patterns. A *Position Specific Scoring Matrix* is a kind of profile that uses a matrix, where each **column** contains information about the same columns of the aforementioned sequences, and each **row** matches one amino acid. In this manner, each **cell** points to a specific column of all observed sequences, as well as one amino acid : the value it contains is the frequency of this amino acid within these columns. PSSMs can be used to align sequences with, the result indicating whether that sequence contains a subsequence that's similar to all of the sequences represented by the PSSM.\n",
    "\n",
    "### Implementation\n",
    "\n",
    "The eponym class `PSSM` contains the frequency matrix as well as metadata that will allow us to use it efficiently as a scoring system when aligning sequences with it (more on that later). Here is the class, as well as a dictionary of absolute amino acid frequencies within the UniProt database (also used to provide scores) :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# AA frequencies for complete UniProt database\n",
    "# from http://web.expasy.org/docs/relnotes/relstat.html, \"AMINO ACID COMPOSITION\"\n",
    "uniprob = {\n",
    "\tAminoAcid(\"Ala\") : .0826,\n",
    "\tAminoAcid(\"Gln\") : .0393,\n",
    "\tAminoAcid(\"Leu\") : .0965,\n",
    "\tAminoAcid(\"Ser\") : .0660,\n",
    "\tAminoAcid(\"Arg\") : .0553,\n",
    "\tAminoAcid(\"Glu\") : .0674,\n",
    "\tAminoAcid(\"Lys\") : .0582,\n",
    "\tAminoAcid(\"Thr\") : .0535,\n",
    "\tAminoAcid(\"Asn\") : .0406,\n",
    "\tAminoAcid(\"Gly\") : .0708,\n",
    "\tAminoAcid(\"Met\") : .0241,\n",
    "\tAminoAcid(\"Trp\") : .0109,\n",
    "\tAminoAcid(\"Asp\") : .0546,\n",
    "\tAminoAcid(\"His\") : .0227,\n",
    "\tAminoAcid(\"Phe\") : .0386,\n",
    "\tAminoAcid(\"Tyr\") : .0292,\n",
    "\tAminoAcid(\"Cys\") : .0137,\n",
    "\tAminoAcid(\"Ile\") : .0593,\n",
    "\tAminoAcid(\"Pro\") : .0472,\n",
    "\tAminoAcid(\"Val\") : .0687,\n",
    "\t\n",
    "}\n",
    "\n",
    "\n",
    "class PSSM:\n",
    "\t\"\"\"\n",
    "\tPosition Specific Score Matrix.\n",
    "\tCreates a profile for a series of aligned sequences, and gives a score to each AA subsitution in a given column.\n",
    "\t\"\"\"\n",
    "\tdef __init__(self, description=\"\"):\n",
    "\t\tself.description=description\n",
    "\t\tself.seqCount = 0 #total number of sequences\n",
    "\t\tself.size = None #all sequences have the same size\n",
    "\t\tself.aaDistribution = None #amino acid distribution\n",
    "\t\tself.aaCount = None\n",
    "\t\tself.gapPenalties = None\n",
    "\t\t\n",
    "\t\n",
    "\tdef add(self, sequence):\n",
    "\t\t#check sequence size\n",
    "\t\tif self.size is None:\n",
    "\t\t\tself.size = len(sequence)\n",
    "\t\t\tself.aaDistribution = [{} for i in range(self.size)]\n",
    "\t\t\tself.aaCount = [0 for i in range(self.size)]\n",
    "\t\t\tself.gapPenalties = [0 for i in range(self.size + 1)]\n",
    "\t\t\n",
    "\t\tassert(len(sequence) == self.size)\n",
    "\t\t\t\n",
    "\t\t#update amino acid count for each column\n",
    "\t\tfor index in range(self.size):\n",
    "\t\t\tif not sequence[index].isGap():\n",
    "\t\t\t\tself.aaCount[index] += 1\n",
    "\t\t\t\ttry:\n",
    "\t\t\t\t\tself.aaDistribution[index][sequence[index]] += 1\n",
    "\t\t\t\texcept:\n",
    "\t\t\t\t\tself.aaDistribution[index][sequence[index]] = 1\n",
    "\t\t\n",
    "\t\t#increase sequence count\n",
    "\t\tself.seqCount += 1\n",
    "\t\t\n",
    "\tdef getDescription(self):\n",
    "\t\treturn self.description\n",
    "\t\t\n",
    "\tdef getScore(self, aminoAcid, columnIndex):\n",
    "\t\t#pseudocounts\n",
    "\t\talpha = self.aaCount[columnIndex] - 1\n",
    "\t\tbeta = sqrt(self.seqCount)\n",
    "\t\talphaplusbeta = alpha + beta\n",
    "\n",
    "\t\t#random probability of amino acid\n",
    "\t\ttry:\n",
    "\t\t\tp_aa = uniprob[aminoAcid]\n",
    "\t\texcept:\n",
    "\t\t\tp_aa = 0.001\n",
    "\t\t\n",
    "\t\t#evolutionary probability of amino acid\n",
    "\t\ttry:\n",
    "\t\t\tf_aa = self.aaDistribution[columnIndex][aminoAcid] / self.seqCount\n",
    "\t\texcept:\n",
    "\t\t\tf_aa = 0\n",
    "\t\t\t\n",
    "\t\tq_aa = (alpha * f_aa + beta * p_aa) / alphaplusbeta\n",
    "\t\t\n",
    "\t\treturn log(q_aa / p_aa)\n",
    "\t\n",
    "\t\n",
    "\tdef getGapPenalty(self, columnIndex):\n",
    "\t\treturn self.gapPenalties[columnIndex]\n",
    "\t\n",
    "\t\n",
    "\tdef setGapPenalty(self, penalty, columnIndex=None):\n",
    "\t\tif columnIndex is None:\n",
    "\t\t\tfor i in range(self.size):\n",
    "\t\t\t\tself.gapPenalties[i] = penalty\n",
    "\t\telse:\n",
    "\t\t\tself.gapPenalties[columnIndex] = penalty\n",
    "\t\n",
    "\tdef __len__(self):\n",
    "\t\treturn self.size\n",
    "\t\n",
    "\tdef __repr__(self):\n",
    "\t\tfor i in range(self.size):\n",
    "\t\t\tfor key, score in self.aaDistribution[i].items():\n",
    "\t\t\t\tprint(key, \": \", score, \"(\", self.getScore(key, i), \")\", sep=\"\",  end=\", \")\n",
    "\t\t\tprint()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scoring\n",
    "\n",
    "Suppose we want to align a sequence to a PSSM: it will be the PSSM that determines the scores for the alignment -based on the frequencies gathered from previously aligned sequences-. Sure, PSSM has \"Scoring Matrix\" in it, but we haven't talked about scores yet (that is, unless you've read the code above). That's when you should think : \"oh, but I know what a scoring matrix is, it gives out scores for each amino acid pair, that's what this is\". Well, when we were aligning two sequences together, that was the case since we made the assumption that the positions of amino acids didn't matter. However the goal now is to align one sequence with a load of other aligned sequences -represented by a PSSM- that often don't have the same amino acid at the same position. The frequency and range of amino acids can vary for each column of the PSSM, therefore the column value is required (instead of the \"other\" amino acid) to provide a score.\n",
    "\n",
    "So how is that score calculated you ask ? Well, exactly the same way we did for the Blosum matrices: with the **log-odds ratio** $\\frac{q_{a,b}}{p_a}$, where $q_{a,c}$ is the evolutionary probability of amino acid $a$ being located at column $c$, and $p_a$ is the random probability of amino acid $a$ (which doesn't depend on the column). What differs from the previous project is the way these terms are calculated:\n",
    "* The **random probability** is based on the frequencies of amino acids in the whole UniProt database, which can be found [here](http://web.expasy.org/docs/relnotes/relstat.html).\n",
    "* The **evolutionary probability**\n",
    "\n",
    "Protein [D6C652](http://www.uniprot.org/uniprot/D6C652) (Transcriptional coactivator YAP1-A)\n",
    "![title](D6C652-Domains.PNG)\n",
    "\n",
    "Protein [P46935](http://www.uniprot.org/uniprot/P46935) (E3 ubiquitin-protein ligase NEDD4)\n",
    "![title](P46935-Domains.PNG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " >>> Creating PSSM from file msaresults-MUSCLE.fasta ... done\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 59\n",
      "Type       : local\n",
      "Score      : 25.92\n",
      "Gaps       : 28\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t28 Gaps, 31 AAs (positions 142 to 173)\n",
      "\n",
      "142\n",
      "-LPPGWEMAKT-PS-GQR-YFLN------------------------HIDQTTTWQDPR\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 60\n",
      "Type       : local-suboptimal(1)\n",
      "Score      : 23.73\n",
      "Gaps       : 28\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t28 Gaps, 32 AAs (positions 200 to 232)\n",
      "\n",
      "200\n",
      "-LPDGWEQALTPEGEA---YFIN------------------------HKNKSTSWLDPRL\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 5\n",
      "Type       : local-suboptimal(2)\n",
      "Score      : 11.06\n",
      "Gaps       : 1\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t1 Gaps, 4 AAs (positions 254 to 258)\n",
      "\n",
      "254\n",
      "-LPPP\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 60\n",
      "Type       : local-suboptimal(3)\n",
      "Score      : 10.73\n",
      "Gaps       : 41\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t41 Gaps, 19 AAs (positions 121 to 140)\n",
      "\n",
      "121\n",
      "-LA-P-------PSA--P-----------------------------HL-RQSSYEIPDD\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 60\n",
      "Type       : local\n",
      "Score      : 27.34\n",
      "Gaps       : 28\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|P46935|NEDD4_MOUSE E3 ubiquitin-protein ligase NEDD4 OS=Mus musculus GN=Nedd4 PE=1 SV=3\n",
      "\t28 Gaps, 32 AAs (positions 461 to 493)\n",
      "\n",
      "461\n",
      "-LPPGWEER-T--H---T---DGRVFFIN------------------HNIKKTQWEDPRL\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 60\n",
      "Type       : local-suboptimal(1)\n",
      "Score      : 25.16\n",
      "Gaps       : 28\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|P46935|NEDD4_MOUSE E3 ubiquitin-protein ligase NEDD4 OS=Mus musculus GN=Nedd4 PE=1 SV=3\n",
      "\t28 Gaps, 32 AAs (positions 406 to 438)\n",
      "\n",
      "406\n",
      "-LPPGW-EEK------Q----DDRGRSYYVD----------------HNSKTTTWSKPTM\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 11\n",
      "Type       : local-suboptimal(2)\n",
      "Score      : 24.98\n",
      "Gaps       : 1\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|P46935|NEDD4_MOUSE E3 ubiquitin-protein ligase NEDD4 OS=Mus musculus GN=Nedd4 PE=1 SV=3\n",
      "\t1 Gaps, 10 AAs (positions 250 to 260)\n",
      "\n",
      "250\n",
      "-LPPGWEERQD\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 56\n",
      "Type       : local-suboptimal(3)\n",
      "Score      : 23.58\n",
      "Gaps       : 27\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|P46935|NEDD4_MOUSE E3 ubiquitin-protein ligase NEDD4 OS=Mus musculus GN=Nedd4 PE=1 SV=3\n",
      "\t27 Gaps, 29 AAs (positions 251 to 280)\n",
      "\n",
      "251\n",
      "-PP---GW--EE--RQ---DVLGRTYYVN----------------HESRRTQWKRP\n",
      "\n",
      "\n",
      "\n",
      " >>> Creating PSSM from file msaresults-OMEGA.fasta ... done\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 13\n",
      "Type       : local\n",
      "Score      : 25.83\n",
      "Gaps       : 1\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t1 Gaps, 12 AAs (positions 142 to 154)\n",
      "\n",
      "142\n",
      "-LPPGWEMAKTPS\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 46\n",
      "Type       : local-suboptimal(1)\n",
      "Score      : 22.23\n",
      "Gaps       : 26\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t26 Gaps, 20 AAs (positions 153 to 173)\n",
      "\n",
      "153\n",
      "-SGQRYFLNHID-------------------------QTTTWQDPR\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 12\n",
      "Type       : local-suboptimal(2)\n",
      "Score      : 20.5\n",
      "Gaps       : 1\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t1 Gaps, 11 AAs (positions 200 to 211)\n",
      "\n",
      "200\n",
      "-LPDGWEQALTP\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 59\n",
      "Type       : local-suboptimal(3)\n",
      "Score      : 20.26\n",
      "Gaps       : 35\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|D6C652|YAP1A_XENLA Transcriptional coactivator YAP1-A OS=Xenopus laevis GN=yap1-a PE=1 SV=1\n",
      "\t35 Gaps, 24 AAs (positions 208 to 232)\n",
      "\n",
      "208\n",
      "-L-TP--EGE-A------YF--INHK----------------------NKSTSWLDPRL\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 11\n",
      "Type       : local\n",
      "Score      : 24.98\n",
      "Gaps       : 1\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|P46935|NEDD4_MOUSE E3 ubiquitin-protein ligase NEDD4 OS=Mus musculus GN=Nedd4 PE=1 SV=3\n",
      "\t1 Gaps, 10 AAs (positions 250 to 260)\n",
      "\n",
      "250\n",
      "-LPPGWEERQD\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 23\n",
      "Type       : local-suboptimal(1)\n",
      "Score      : 24.1\n",
      "Gaps       : 2\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|P46935|NEDD4_MOUSE E3 ubiquitin-protein ligase NEDD4 OS=Mus musculus GN=Nedd4 PE=1 SV=3\n",
      "\t2 Gaps, 21 AAs (positions 461 to 482)\n",
      "\n",
      "461\n",
      "-LPPGWEER-THTDGRVFFINHN\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 24\n",
      "Type       : local-suboptimal(2)\n",
      "Score      : 24.02\n",
      "Gaps       : 13\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|P46935|NEDD4_MOUSE E3 ubiquitin-protein ligase NEDD4 OS=Mus musculus GN=Nedd4 PE=1 SV=3\n",
      "\t13 Gaps, 11 AAs (positions 261 to 272)\n",
      "\n",
      "261\n",
      "-L------------GRTYYVNHES\n",
      "\n",
      "\n",
      "---------- Multi-Seq. Alignment ----------\n",
      "Size       : 12\n",
      "Type       : local-suboptimal(3)\n",
      "Score      : 23.48\n",
      "Gaps       : 1\n",
      "\n",
      "PSSM : WW domain\n",
      "Aligned seq. : sp|P46935|NEDD4_MOUSE E3 ubiquitin-protein ligase NEDD4 OS=Mus musculus GN=Nedd4 PE=1 SV=3\n",
      "\t1 Gaps, 11 AAs (positions 406 to 417)\n",
      "\n",
      "406\n",
      "-LPPGWEEKQDD\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for aligned in (r\"msaresults-MUSCLE.fasta\", r\"msaresults-OMEGA.fasta\"):\n",
    "\tprint(\"\\n >>> Creating PSSM from file {} ...\".format(aligned), end=\"\")\n",
    "\tpssm = PSSM(\"WW domain\")\n",
    "\tfor seq in loadFasta(aligned):\n",
    "\t\tpssm.add(seq)\n",
    "\tpssm.setGapPenalty(4)\n",
    "\tprint(\" done\\n\\n\")\n",
    "\t\n",
    "\t\n",
    "\tal = Align(pssm)\n",
    "\tfor toalign in loadFasta(r\"test.fasta\"):\n",
    "\t\tfor aligned in al.multiAlign(toalign):\n",
    "\t\t\tprint(aligned)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
